---
title: "CRISPRscope: comparison"
author: "Thibault Schowing"
date: "04/01/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Comparison between CRISPRscope and CRISPRscope_meta data



# Import packages
```{r}

library(tidyverse)
library(ggpmisc)


library(ggVennDiagram)


# Load library
#install.packages("VennDiagram")
library(VennDiagram) #Caca

# Prepare a palette of 2 colors with R colorbrewer:
library(RColorBrewer)

#Load
library("jpeg")
library("tiff")
library("magick")


```





# Import data
```{r}
# Restore the objects

CRISPRscope_tbl <- readRDS(file = "./OUTPUT/CRISPRscope_tbl.rds")
#CRISPRscope_meta_tbl <- readRDS(file = "./OUTPUT/CRISPRscope_meta_tbl.rds")

CRISPRscope_meta_tbl_filtered <- readRDS(file = "./OUTPUT/CRISPRscope_meta_tbl_filtered.rds")

```
#----------------------
# Genome vs metagenome repeats

# Export repeats ALL (distinct)

To show common repeats between genomes and metagenomes, cluster all together and show Venn diagram of shared clusters between genome and metagenome

```{r, eval=FALSE}
repeats_genome <- CRISPRscope_tbl %>% select(DR_seq) %>% distinct(DR_seq) %>% 
  rowid_to_column("header") %>% mutate(seq = DR_seq) %>% select(header, seq) %>%
  add_column(sufx = "G") %>% 
  unite(header, c(sufx, header))

repeats_metagenome <- CRISPRscope_meta_tbl_filtered %>%
  filter(Coverage > 1) %>%  # already done in quality filtering
  select(repeat_seq) %>% 
  distinct(repeat_seq) %>% 
  rowid_to_column("header") %>% 
  mutate(seq = repeat_seq) %>%
  select(header, seq) %>% 
  add_column(sufx = "M") %>% 
  unite(header, c(sufx, header)) 

outr <- bind_rows(repeats_genome, repeats_metagenome)

dfr <- data.frame(outr$header, outr$seq)
dfr.fasta = dataframe2fas(dfr, file = "./OUTPUT/CRISPRscope_ALL_REPEATS.fasta")


```

### 80% similarity clustering - repeats

```{r}

Clusters_CRISPRscope_ALL_REPEATS <- read_csv("C:/Users/thsch/Desktop/master_thesis_2/1_scripts/INPUT/Clusters_CRISPRscope_ALL_REPEATS.csv")

Clusters_CRISPRscope_ALL_REPEATS <- Clusters_CRISPRscope_ALL_REPEATS %>% select(seqid, cluster)


clst_G <- Clusters_CRISPRscope_ALL_REPEATS %>% filter(grepl("^G_", seqid)) %>% select(cluster) %>% distinct(cluster)
clst_M <- Clusters_CRISPRscope_ALL_REPEATS %>% filter(grepl("^M_", seqid)) %>% select(cluster) %>% distinct(cluster)



x <- list(
  A = length(clst_G$cluster),
  B = length(clst_M$cluster),
  C = length(intersect(clst_G$cluster,clst_M$cluster))
)

grid.newpage()
venn.plot <- draw.pairwise.venn(area1 = x[[1]], area2 = x[[2]], cross.area = x[[3]], category = c("Genome wide repeats", 
    "Metagenome wide repeats"), lty = rep("blank", 2), fill = c("light blue", "light green"), 
    alpha = c(0.4, 0.4), cat.pos = c(18, 180), euler.d = TRUE, sep.dist = 0.03, cex = c(3,4,4), cat.cex = c(2,2),
    rotation.degree = 0)
# Writing to file


tiff(filename = "./IMG/report/12_2_RepeatsMG80percent.tiff", width = 660, height = 660);
grid.draw(venn.plot);
dev.off();

#img <- readTIFF("./IMG/report/SpacerMG100percent.tiff", native=TRUE)
#writeJPEG(img, target = "./IMG/report/SpacerMG100percent.jpeg", quality = 1)

img <- image_read("./IMG/report/12_2_RepeatsMG80percent.tiff")
image_write(img, format = "pdf", "./IMG/report/12_2_RepeatsMG80percent.pdf")

fn <- "./IMG/report/12_2_RepeatsMG80percent.tiff"
if (file.exists(fn)) {
  #Delete file if it exists
  file.remove(fn)
}
```

## <stat>
```{r}
126+23+4
23/27
23/153
```



#----------------------
# Genome vs metagenome spacers

# Export spacers ALL (distinct)

To show the common spacers between the genomes and metagenomes, we cluster alltogether the spacers from the two datasets and instead of creating a Venn diagram of shared sequences, we create a venn diagram of shared clusters. Metagenomic spacers with Coverage < 1 are ommited. Only distinct sequences are taken. 
```{r, eval=FALSE}
spacers_genome <- CRISPRscope_tbl %>% select(SpacerSeq) %>% distinct(SpacerSeq) %>% 
  rowid_to_column("header") %>% mutate(seq = SpacerSeq) %>% select(header, seq) %>%
  add_column(sufx = "G") %>% 
  unite(header, c(sufx, header))

spacers_metagenome <- CRISPRscope_meta_tbl_filtered %>%
  filter(Coverage > 1) %>%  # already done in quality fitering
  select(spacer_seq) %>% 
  distinct(spacer_seq) %>% 
  rowid_to_column("header") %>% 
  mutate(seq = spacer_seq) %>%
  select(header, seq) %>% 
  add_column(sufx = "M") %>% 
  unite(header, c(sufx, header)) 

out2 <- bind_rows(spacers_genome, spacers_metagenome)

# Total number of spacers
out2 %>% distinct(seq)


dfs <- data.frame(out2$header, out2$seq)
dfs.fasta = dataframe2fas(dfs, file = "./OUTPUT/CRISPRscope_ALL_SPACERS.fasta")


```

##Load all spacers clusters
Now, load the cd-hit-est result and check it out !
### 80% similarity clustering - spacers
```{r }
Clusters_CRISPRscope_ALL_SPACERS <- read_csv("C:/Users/thsch/Desktop/master_thesis_2/1_scripts/INPUT/Clusters_CRISPRscope_ALL_SPACERS.csv")

Clusters_CRISPRscope_ALL_SPACERS <- Clusters_CRISPRscope_ALL_SPACERS %>% select(seqid, cluster)


clst_G <- Clusters_CRISPRscope_ALL_SPACERS %>% filter(grepl("^G_", seqid)) %>% select(cluster) %>% distinct(cluster)
clst_M <- Clusters_CRISPRscope_ALL_SPACERS %>% filter(grepl("^M_", seqid)) %>% select(cluster) %>% distinct(cluster)



x <- list(
  A = length(clst_G$cluster),
  B = length(clst_M$cluster),
  C = length(intersect(clst_G$cluster,clst_M$cluster))
)

grid.newpage()
venn.plot <- draw.pairwise.venn(area1 = x[[1]], area2 = x[[2]], cross.area = x[[3]], category = c("Genome wide spacers", 
    "Metagenome wide spacers"), lty = rep("blank", 2), fill = c("light blue", "light green"), 
    alpha = c(0.4, 0.4), cat.pos = c(0, 180), euler.d = TRUE, sep.dist = 0.03, cex = c(4,3,4), cat.cex = c(2,2),
    rotation.degree = 0)
# Writing to file


tiff(filename = "./IMG/report/12_1_SpacerMG80percent.tiff", width = 660, height = 660);
grid.draw(venn.plot);
dev.off();

#img <- readTIFF("./IMG/report/SpacerMG100percent.tiff", native=TRUE)
#writeJPEG(img, target = "./IMG/report/SpacerMG100percent.jpeg", quality = 1)

img <- image_read("./IMG/report/12_1_SpacerMG80percent.tiff")
image_write(img, format = "pdf", "./IMG/report/12_1_SpacerMG80percent.pdf")

fn <- "./IMG/report/12_1_SpacerMG80percent.tiff"
if (file.exists(fn)) {
  #Delete file if it exists
  file.remove(fn)
}

```

Compare genome vs metagenome spacers -> venn diagram 

```{r}
unique_genome_spacers <- CRISPRscope_tbl %>% select(SpacerSeq) %>% distinct(SpacerSeq)
unique_metagenome_spacers <- CRISPRscope_meta_tbl_filtered %>% select(spacer_seq) %>% distinct(spacer_seq)

unique_genome_repeats <- CRISPRscope_tbl %>% select(DR_seq) %>% distinct(DR_seq)
unique_metagenome_repeats <- CRISPRscope_meta_tbl_filtered %>% select(repeat_seq) %>% distinct(repeat_seq)

```


## Venn diagram without clustering


```{r}
x <- list(
  A = length(unique_genome_spacers$SpacerSeq),
  B = length(unique_metagenome_spacers$spacer_seq),
  C = length(intersect(unique_genome_spacers$SpacerSeq,unique_metagenome_spacers$spacer_seq))
)
# ggVennDiagram(x)+
#   ggplot2::scale_fill_gradient(low="blue",high = "yellow") 

grid.newpage()
draw.pairwise.venn(area1 = x[[1]], area2 = x[[2]], cross.area = x[[3]], category = c("Genome wide spacers", 
    "Metagenome wide spacers"), lty = rep("blank", 2), fill = c("light blue", "light green"), 
    alpha = rep(0.5, 2), cat.pos = c(0, 180), euler.d = TRUE, sep.dist = 0.03, 
    rotation.degree = 0)



# Now for the metagenome: we take only reads with coverage > 1 and we take the distinct clusters to avoid similar spacers. 


unique_metagenome_spacers_cluster <- CRISPRscope_meta_tbl_filtered %>% 
  filter(Coverage > 1) %>% 
  distinct(cluster_spacer)


y <- list(
  A = length(unique_genome_spacers$SpacerSeq),
  B = length(unique_metagenome_spacers$spacer_seq),
  C = length(intersect(unique_genome_spacers$SpacerSeq,unique_metagenome_spacers$spacer_seq))
)
# ggVennDiagram(x)+
#   ggplot2::scale_fill_gradient(low="blue",high = "yellow") 

grid.newpage()
draw.pairwise.venn(area1 = x[[1]], area2 = x[[2]], cross.area = x[[3]], category = c("Genome wide spacers", 
    "Metagenome wide spacers"), lty = rep("blank", 2), fill = c("light blue", "light green"), 
    alpha = rep(0.5, 2), cat.pos = c(0, 180), euler.d = TRUE, sep.dist = 0.03, 
    rotation.degree = 0)

```




### Old shit

Diagram includes genomic and metagenomic spacers (or repeats) found in the project. 
```{r}
myCol <- c("blue","red")#brewer.pal(2, "Pastel2")


# Chart
venn.diagram(
  x = list(unique_genome_spacers$SpacerSeq, unique_metagenome_spacers$spacer_seq),
  category.names = c("Genome spacers" , "Metagenome spacers " ),
  filename = './IMG/metagenome/COMP_Shared_spacers_venn_diagramm.png',
  output=F,
  # Output features
  imagetype="png" ,
  height = 480 , 
  width = 480 , 
  resolution = 300,
  
  # Circles
  lwd = 2,
  lty = 'blank',
  fill = myCol,
  
  # Numbers
  cex = .5,
  fontfamily = "sans",
  
  # Set names
  cat.cex = 0.4,
  cat.default.pos = "outer",
  cat.pos = c(0, 0),
  cat.dist = c(0, 0),
  cat.fontfamily = "sans"
  
)

```
Repeats
```{r}
myCol <- c("blue","red")#brewer.pal(2, "Pastel2")


# Chart
venn.diagram(
  x = list(unique_genome_repeats$DR_seq, unique_metagenome_repeats$repeat_seq),
  category.names = c("Genome repeats" , "Metagenome repeats " ),
  filename = './IMG/metagenome/COMP_Shared_repeats_venn_diagramm.png',
  output=F,
  # Output features
  imagetype="png" ,
  height = 480 , 
  width = 490 , 
  resolution = 300,
  
  # Circles
  lwd = 2,
  lty = 'blank',
  fill = myCol,
  
  # Numbers
  cex = .5,
  fontfamily = "sans",
  
  # Set names
  cat.cex = 0.4,
  cat.default.pos = "outer",
  cat.pos = c(0, 0),
  cat.dist = c(0, 0),
  cat.fontfamily = "sans"
  
)
```

Try the same with only one circle
```{r}
myCol <- c("blue")#brewer.pal(2, "Pastel2")


# Chart
venn.diagram(
  x = list(unique_genome_spacers$SpacerSeq),
  category.names = c("Genome spacers" ),
  filename = './IMG/metagenome/COMP_genome_only_spacers_venn_diagramm.png',
  output=F,
  # Output features
  imagetype="png" ,
  height = 480 , 
  width = 480 , 
  resolution = 300,
  
  # Circles
  lwd = 2,
  lty = 'blank',
  fill = myCol,
  
  # Numbers
  cex = .5,
  fontfamily = "sans",
  
  # Set names
  cat.cex = 0.4,
  cat.default.pos = "outer",
  cat.pos = c(-0.5),
  cat.dist = c(-0.05),
  cat.fontfamily = "sans"
  
)
```
repeats
```{r}
myCol <- c("blue")#brewer.pal(2, "Pastel2")


# Chart
venn.diagram(
  x = list(unique_genome_repeats$DR_seq),
  category.names = c("Genome repeats" ),
  filename = './IMG/metagenome/COMP_genome_only_repeats_venn_diagramm.png',
  output=F,
  # Output features
  imagetype="png" ,
  height = 480 , 
  width = 480 , 
  resolution = 300,
  
  # Circles
  lwd = 2,
  lty = 'blank',
  fill = myCol,
  
  # Numbers
  cex = .5,
  fontfamily = "sans",
  
  # Set names
  cat.cex = 0.4,
  cat.default.pos = "outer",
  cat.pos = c(-0.5),
  cat.dist = c(-0.05),
  cat.fontfamily = "sans"
  
)
```

## barplot

## <p>
```{r fig.height=8, fig.width=12}
unique_genome_spacers <- CRISPRscope_tbl %>% select(SpacerSeq) %>% distinct(SpacerSeq)
unique_metagenome_spacers <- CRISPRscope_meta_tbl_filtered %>% select(spacer_seq) %>% distinct(spacer_seq)

unique_genome_repeats <- CRISPRscope_tbl %>% select(DR_seq) %>% distinct(DR_seq)
unique_metagenome_repeats <- CRISPRscope_meta_tbl_filtered %>% select(repeat_seq) %>% distinct(repeat_seq)

u_g_sp <- length(unique_genome_spacers$SpacerSeq)
u_g_dr <- length(unique_genome_repeats$DR_seq)

u_m_sp <- length(unique_metagenome_spacers$spacer_seq)
u_m_dr <- length(unique_metagenome_repeats$repeat_seq)



df <- tibble(
  group = c("Genome spacers", "Genome repeats", "Metagenome spacers", "Metagenome repeats"),
  value = c(u_g_sp, u_g_dr, u_m_sp, u_m_dr)
)

plt2 <- df %>% ggplot(aes(x = group, y = value, fill = group, label = value)) +
  geom_bar(position="dodge", stat="identity") +
  geom_label(size = 3, color = "black",fill="white") + 
  xlab("") +
  ylab("Sequences count")+
  #ggtitle("Number of Spacers and Repeats detected in the genome assemblies and in the metagenomes.") + 
  scale_fill_manual(values=c("green", "darkgreen", "lightblue", "blue"), labels = c("Genome wide repeats", "Genome wide spacers", "Metagenome wide repeats", "Metagenome wide spacers")) + 
  guides(fill=guide_legend(title="")) +
  #geom_text(aes(label=nb_strain), position=position_dodge(width=0.9), color = "lightblue", vjust=-1) +
  #geom_text(aes(label=CRISPR_assembly_ratio), position=position_dodge(width=0.9), color = "lightblue", vjust=-1) +
  theme_bw() +
  theme(
    axis.text.x = element_text(
      angle = 45,
      hjust = 1,
      color = "black"
    ),
    axis.text.y = element_text(color = "black"),
    axis.title.x = element_text(color = "black"),
    axis.title.y = element_text(color = "black"),
    axis.line.y.left = element_line(color = "black"),
    axis.line.y.right = element_line(color = "black"),
    axis.line.x = element_line(color = "black"),
    axis.ticks = element_line(color = "black"),
    plot.title = element_text(color = "black"),
    panel.background = element_rect(fill = "white", colour = NA),
    plot.background = element_rect(fill = "white", colour = NA),
    legend.position = "top",
    legend.background = element_rect(fill = "white", colour = NA),
    legend.box.background = element_rect(fill = "white", colour = NA),
    legend.text = element_text(color = "black"),
    legend.title = element_text(color = "black"),
    plot.margin = unit(c(1, 1, 1, 1), "cm"),
    panel.grid.major = element_line(color = "black", size = 0.1),
    panel.grid.minor = element_line(color = "black", size = 0.1),
    strip.background = element_rect(
      color = "black",
      fill = "white",
      size = 0.5,
      linetype = "solid"
    ),
    strip.text = element_text(color = "black")
  )

plt2

ggsave(
  plot = plt2,
  file = "./IMG/report/SPDRcountComparison.pdf",
  bg = "white",
  width = 25,
  height = 15,
  units = "cm",
  dpi = 800
)


```














#------------------------------

# Common repeats/spacers between projects


```{r}
sub_gen_sp <- CRISPRscope_tbl %>% select(Organism, Strain, SpacerSeq) %>% distinct(Organism, Strain, SpacerSeq)
sub_metagen_sp <- CRISPRscope_meta_tbl_filtered %>% select(ProjectID, SRA_ID, GID, SPID, spacer_seq) %>% distinct(ProjectID, SRA_ID, GID, SPID, spacer_seq)

sub_gen_dr <- CRISPRscope_tbl %>% select(Organism, Strain, DR_seq) %>% distinct(Organism, Strain, DR_seq)
sub_metagen_dr <- CRISPRscope_meta_tbl_filtered %>% select(ProjectID, SRA_ID, GID, repeat_seq) %>% distinct(ProjectID, SRA_ID, GID, repeat_seq)



unique_genome_repeats <- CRISPRscope_tbl %>% select(Organism, Strain, DR_seq) %>% distinct(Organism, Strain, DR_seq)
unique_metagenome_repeats <- CRISPRscope_meta_tbl_filtered %>% select(ProjectID, SRA_ID, GID, SPID, repeat_seq) %>% distinct(ProjectID, SRA_ID, GID, SPID, repeat_seq)


# --------- Spacers ----------
shared_spacers <- inner_join(sub_gen_sp, sub_metagen_sp, by = c("SpacerSeq" = "spacer_seq"), copy = FALSE)

shared_spacers %>% group_by(Organism, ProjectID) %>% summarise(nb_match_spacer_in_meta = n())


# --------- Repeats ----------

shared_repeats <- inner_join(sub_gen_dr, sub_metagen_dr, by = c("DR_seq"="repeat_seq"), copy = FALSE)

shared_repeats %>% group_by(Organism, ProjectID) %>% summarise(nb_match_repeat_in_meta = n())

```



#------------------------------
# BLAST

Export all (genome and metagenome) cluster reference for blasting

## <data> Import merge and filter
```{r}

# Cluster, sequence ID and sequence 
Clusters_CRISPRscope_ALL_SPACERS <- read_csv("C:/Users/thsch/Desktop/master_thesis_2/1_scripts/INPUT/Clusters_CRISPRscope_ALL_SPACERS.csv")

out2 <- Clusters_CRISPRscope_ALL_SPACERS %>% arrange(cluster) %>% filter(identity == 1) %>%
  mutate(header = cluster, seq = seq) %>%
  select(header, seq)

dfs <- data.frame(out2$header, out2$seq)
dfs.fasta = dataframe2fas(dfs, file = "./OUTPUT/cluster_reference_CRISPRscope.fasta")






```


##<stat> count hits

```{r}

# Use Clusters_CRISPRscope_ALL_SPACERS to get the sequence (cluster ref) and the seq_id
cluster_references <- Clusters_CRISPRscope_ALL_SPACERS %>% arrange(cluster) %>% filter(identity == 1)

final_spacer_hits <- read_csv("C:/Users/thsch/Desktop/master_thesis_2/1_scripts/INPUT/final_spacer_hits.txt")

BLAST_results <- final_spacer_hits %>% left_join(cluster_references, by = c("spacer" = "cluster")) %>% 
  select(-X1)


# Separate genomic vs metagenomic results

BLAST_results <- BLAST_results %>% separate(seqid, c("Origin", "ID"), '_') 

# Genome
G <- BLAST_results %>% filter(Origin == "G") %>%  count(final)
g_total <- sum(G$n) # 6247
g_bacteria <- G$n[[1]]
g_nohit <- G$n[[2]]
g_phage <- G$n[[3]]

g_bacteria/g_total # 18.8 %
g_nohit/g_total    # 42.4 %
g_phage/g_total    # 38.7 %

# Metagenome
M <- BLAST_results %>% filter(Origin == "M") %>%  count(final)
m_total <- sum(M$n) # 5731
m_bacteria <- M$n[[1]]
m_nohit <- M$n[[2]]
m_phage <- M$n[[3]]

m_bacteria/m_total # 11.4 %
m_nohit/m_total    # 52.9 %
m_phage/m_total    # 35.5 %

#Overall

Tot <- BLAST_results %>% count(final) 

t_total <- sum(Tot$n)
t_bacteria <- Tot$n[[1]]
t_nohit <- Tot$n[[2]]
t_phage <- Tot$n[[3]]

t_bacteria/t_total # 15.3 %
t_nohit/t_total    # 47.4 %
t_phage/t_total    # 37.2 %


```

## <p>

```{r fig.height=6, fig.width=8}


# Genome
df_G <- tibble(
  group = c("Phage", "Bacterial", "No hit"),
  value = c(2418, 1177, 2652)
)

plot_blast_G <- df_G %>% ggplot(aes(x=group, y=value)) +
  geom_segment( aes(x=group, xend=group, y=0, yend=value), color="lightblue3", size=2) +
  geom_point(size=2, color="blue", alpha=0.7, shape=18, stroke=2) +
  geom_text(aes(label=value),hjust=1.4, vjust=0.2) +
  #xlab("Genomic spacers") +
  xlab("")+
  ylab("") +
  ylim(c(0,3050)) +
  theme_bw() +
  theme(
    axis.text.x = element_text(
      angle = 49,
      hjust = 0.9,
      color = "black",
      size = 12
    ),
    legend.position = "none",
    plot.margin = unit(c(1, 1, 1, 1), "cm"),
    panel.grid.major = element_line(color = "black", size = 0.1),
    panel.grid.minor = element_line(color = "black", size = 0.1),
    strip.background = element_rect(
      color = "black",
      fill = "white",
      size = 0.5,
      linetype = "solid"
    ),
    strip.text = element_text(color = "black")
  )

# Metagenome
df_M <- tibble(
  group = c("Phage", "Bacterial", "No hit"),
  value = c(2039, 658, 3034)
)

plot_blast_M <- df_M %>% ggplot(aes(x=group, y=value)) +
  geom_segment( aes(x=group, xend=group, y=0, yend=value), color="lightblue3", size=2) +
  geom_point( size=2, color="blue", alpha=0.7, shape=18, stroke=2) +
  geom_text(aes(label=value),hjust=1.4, vjust=0.2) +
  #xlab("Metagenomic spacers") +
  xlab("")+
  ylab("") +
  ylim(c(0,3050)) +
  theme_bw() +
  theme(
    axis.text.x = element_text(
      angle = 49,
      hjust = 0.9,
      color = "black",
      size = 12
    ),
    legend.position = "none",
    plot.margin = unit(c(1, 1, 1, 1), "cm"),
    panel.grid.major = element_line(color = "black", size = 0.1),
    panel.grid.minor = element_line(color = "black", size = 0.1),
    strip.background = element_rect(
      color = "black",
      fill = "white",
      size = 0.5,
      linetype = "solid"
    ),
    strip.text = element_text(color = "black", size = 12)
  )


plt <- ggarrange(plot_blast_G, plot_blast_M, 
          labels = c(" ", " "),
          ncol = 2, nrow = 1)



ggsave(
  plot = plot_blast_G,
  file = "./IMG/report/14_1BLASTn_results_genomic.pdf",
  bg = "white",
  width = 10,
  height = 10,
  units = "cm",
  dpi = 800
)

plot_blast_G


ggsave(
  plot = plot_blast_M,
  file = "./IMG/report/14_2BLASTn_results_metagenomic.pdf",
  bg = "white",
  width = 10,
  height = 10,
  units = "cm",
  dpi = 800
)
plot_blast_M

ggsave(
  plot = plt,
  file = "./IMG/report/14_BLASTn_results.pdf",
  bg = "white",
  width = 20,
  height = 10,
  units = "cm",
  dpi = 800
)

ggsave(
  plot = plt,
  file = "./IMG/poster/14_BLASTn_results.pdf",
  bg = "white",
  width = 20,
  height = 10,
  units = "cm",
  dpi = 800
)

plt


```



##<stat>










































