---
title: "Crass_analysis.rmd"
author: "Thibault Schowing"
date: "26/10/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Analysis of the CRASS outputs


For each metagenome assembly with CRASS, various outputs are produced. 
- spacers.fasta
- DR.fasta
- flankers.fasta
- stats.csv

The stats.csv gives useful informations on the general quantity of CRISPR arrays and their sizes. 


# Various functions

## Imports
```{r ,results='hide', echo=FALSE}
library(readr)
library(tidyverse)
library(plotly)

#remotes::install_github("wilkelab/ggridges")
library(ggridges)

#devtools::install_github('Mikata-Project/ggthemr')
library(ggthemr)

#library(reticulate)


# if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
# BiocManager::install("Biostrings")

library(Biostrings)

# Converts dataframe to fasta file
#install.packages("seqRFLP")
library("seqRFLP")

```






## Utils
```{r ,results='hide', echo=FALSE}


# Get the methods available on DNAStringSet
#methods(class = "DNAStringSet")


#
# Converts dna string set (dss) to dataframe 
#
dss2df <- function(dss) data.frame(length=width(dss), seq=as.character(dss), names=names(dss))

#
# Converts dna string set (dss) to tibble
#
dss2tibble <- function(dss) as_tibble(data.frame(length=width(dss), seq=as.character(dss), names=names(dss)))


```

# Data getters
## DEPRECATED BIOPROJECT entry outputs 
The sample per sample getter is now merging itself all the data. The groups and spacers IDs are not sanitized (check CRISPRtools sanitize function) but they are uniquely identifiable by Bioproject/SRA_ID and then Group and Spacer. 

```{r,results='hide', echo=FALSE}
#DEPRECATED
# get_bioproject_tibbles <- function(projectname, basefolder){
#   
#   #
#   # Spacers
#   #
#   spacers <- paste(basefolder, projectname, "/", projectname , ".Spacers.fasta", sep = "")
#   #spacers <- "C:/Users/thsch/Desktop/master_thesis/4_CRASS/PRJEB6952/PRJEB6952.Spacers.fasta"
#   dna <- readDNAStringSet(spacers)
#   spacer_tbl <- dss2tibble(dna)
#   
#   #
#   # DR
#   #
#   DR <- paste(basefolder, projectname, "/", projectname , ".DR.fasta", sep = "")
#   #dr <- "C:/Users/thsch/Desktop/master_thesis/4_CRASS/PRJEB6952/PRJEB6952.Spacers.fasta"
#   dnadr <- readDNAStringSet(DR)
#   dr_tbl <- dss2tibble(dnadr)
#   
#   
#   #
#   # Stats
#   #
#   
#   # Path 
#   statfile <- paste(basefolder, projectname, "/", projectname , ".stats.csv", sep = "")
#   
#   # Read file
#   stats_tbl <- read_delim(statfile, "\t", escape_double = FALSE, trim_ws = TRUE)
#   
#   
#   return(list("spacers" = spacer_tbl, "dr" = dr_tbl, "stats" = stats_tbl))
# }
```




## Get SRA entries

```{r,results='hide', echo=FALSE}
#
# Gets all CRASS data, sample by sample, adding information as project name, readcount, etc. 
#
get_sra_tibbles <- function(projectname, basefolder){
  

  
  
  # for a bioproject generate 3 tibbles: Spacers, DR and stats. Attention, CRISPRtools sanitized the names for the merged spacers (they are unique) but here they do collide. As the spacer's name is only a place in a graph, we won't care about it now. 
  
  # Spacers
  # - DF with Length / seq / names / SRA ID / read count 
  # 
  # DR
  # - DF with Length / seq / names / SRA ID / read count
  #
  # Stats
  # - Consensus repeat / numbers numbers numbers / SRA ID / read count
  
  
  #spacer_tbl_colnames <- c("length", "seq", "names", "SRA_ID", "reads_count")
  spacer_tbl <- NULL
  dr_tbl <- NULL
  stats_tbl <- NULL
  id_tbl <- NULL # just all the sra entries 
  
  
  # FUnction to generate a big tibble with spacers and SRR id / readcount
  
  SRA_DIRECTORIES <- list.dirs(path=paste(basefolder, projectname, sep = ""), recursive=FALSE)
  
  for (sradir in SRA_DIRECTORIES){
    
    #print(sradir)
    
    #
    # ID
    #
    SRA_ID <- strsplit(sradir[[1]], "/")[[1]][4]
    
    #
    # Read count
    #
    count_file_path <- file.path(paste(sradir, "/", projectname, ".", SRA_ID, ".readcount.txt", sep = ""))
    #print(count_file_path)
    
    
    readcount <- as.numeric(readChar(count_file_path, file.info(count_file_path)$size))
    #print(readcount)
    
    #
    # Spacers (fasta to tibble)
    #
    spacer_file_path <- file.path(paste(sradir, "/", projectname, ".", SRA_ID, ".Spacers.fasta", sep = ""))
    
    tmp_fasta_tibble <- dss2tibble(readDNAStringSet(spacer_file_path))
    
    tmp_fasta_tibble <- tmp_fasta_tibble %>% 
      mutate(SRA_ID = SRA_ID, reads_count = readcount)
    
    #print(length(tmp_fasta_tibble$seq))
    
    spacer_tbl = bind_rows(spacer_tbl, tmp_fasta_tibble)
    
    
    #
    # DR (fasta to tibble)
    #
    dr_file_path <- file.path(paste(sradir, "/", projectname, ".", SRA_ID, ".DR.fasta", sep = ""))
    
    tmp_fasta_tibble <- dss2tibble(readDNAStringSet(dr_file_path))
    
    tmp_fasta_tibble <- tmp_fasta_tibble %>% 
      mutate(SRA_ID = SRA_ID, reads_count = readcount)
    
    dr_tbl = bind_rows(dr_tbl, tmp_fasta_tibble)
    
    
    
    #
    # Stats (csv) Attention; name is csv but file is actually TSV
    #
    stats_file_path <- file.path(paste(sradir, "/", projectname, ".", SRA_ID, ".stats.csv", sep = ""))
    tmp_tsv_stats_tibble <- read_tsv(stats_file_path, col_names = T)
    
    tmp_tsv_stats_tibble <- tmp_tsv_stats_tibble %>% 
      mutate(SRA_ID = SRA_ID, reads_count = readcount)
    
    stats_tbl = bind_rows(stats_tbl, tmp_tsv_stats_tibble)
    
    #
    # ID tibble: contains the SRA_ID and the readcount
    #
    tmp_id_tbl = tibble(`SRA_ID` = SRA_ID, `reads_count` = readcount)
    id_tbl = bind_rows(id_tbl, tmp_id_tbl) 
    
    
  }# for each sample
  
  
  # Operations on finished tibbles
  
  # id_tbl_empty: contains the SRA_IDs of the samples without CRISPRs
  # id_tbl_crispr: contains the SRA_IDs of the samples with CRISPRs
  # id_tbl_out: contains both 
  
  id_tbl_empty <- id_tbl %>% filter(!SRA_ID %in% stats_tbl$SRA_ID) %>% mutate(no_crispr = TRUE)
  
  id_tbl_crispr <- id_tbl %>% filter(!SRA_ID %in% id_tbl_empty$SRA_ID) %>% mutate(no_crispr = FALSE)
  
  id_tbl_out <- bind_rows(id_tbl_crispr, id_tbl_empty)
  
  # SP
  
  spacer_tbl <- spacer_tbl %>% add_column(ProjectID = projectname, .before = "length") %>% 
    relocate(SRA_ID, .after = ProjectID) %>% 
    mutate(GID = str_extract(names, "G[0-9]+"), .after = SRA_ID) %>% 
    mutate(SPID = str_extract(names, "SP[0-9]+"), .after = GID) %>% 
    mutate(Coverage = str_extract(names, "[0-9]+$"), .after = SPID) %>% 
    mutate(spacer_seq = seq) %>% select(-seq)
     # Note: remame() has bugs with developpment version of rlang.
  
  # DR
  
  dr_tbl <- dr_tbl %>% add_column(ProjectID = projectname, .before = "length") %>% 
    relocate(SRA_ID, .after = ProjectID) %>% 
    mutate(GID = str_extract(names, "G[0-9]+"), .after = SRA_ID) %>%  
    mutate(repeat_seq = seq) %>% select(-seq)
  
  
  
  # Stats
  
  stats_tbl <- stats_tbl %>% add_column(ProjectID = projectname, .before = "GID") %>% 
    relocate(SRA_ID, .after = ProjectID)
  
  
  
  
  id_tbl_out <- id_tbl_out %>% add_column(ProjectID = projectname, .before = "SRA_ID")
  
  # Todo here: for spacer and tbl split names column to have group (and spacer ?)
  
  
  return(list("spacers" = spacer_tbl, "dr" = dr_tbl, "stats" = stats_tbl, "id_tbl" = id_tbl_out))

}


```

## Matched Direct Repeats


```{r,results='hide', echo=FALSE}
# PENDING FOR CRISPRmapDB

get_matched_cas_CCFDB <- function(projectname, basedirectory){
  
  file_matched = paste(basedirectory, projectname, "/", projectname, "_matched_cas.csv", sep = "")
  
  if(file.exists(file_matched)){
    mc <- read_delim(file_matched,";", escape_double = F, trim_ws = T) %>% 
    mutate(ProjectID = projectname)
    return(mc)
  }
  else{
    return(NULL)
  }
  
}

#PRJEB6952_matched_cas <- get_matched_cas_CCFDB("PRJEB6952","../4_CRASS/")

#PRJEB6952_matched_cas <- read_delim("C:/Users/thsch/Desktop/master_thesis/4_CRASS/PRJEB6952/PRJEB6952_matched_cas.csv", ";", escape_double = FALSE, trim_ws = TRUE)
```





# Data analysis functions

## Array size

## TODO: switch geom_density_ridge to Plot_ly. 

```{r,results='hide', echo=FALSE}
# arraySize_tbl_grp_ridge <- function(input_tibble, projectname=""){
#   # check rename(input_tibble, nspacers = `# spacers`)
#   
#   d1 <- input_tibble %>% 
#     group_by(SRA_ID) %>% 
#     count(`# spacers`) %>% 
#     mutate(nspacers = `# spacers`) %>% 
#     select(-`# spacers`)
#   
#   #View(d1)
#   median_nb <- median(d1$nspacers)
#   
#   
#   plt <- d1 %>% 
#     ggplot(aes(y=SRA_ID, 
#                x=nspacers, 
#                fill=SRA_ID, 
#                color=SRA_ID)) +
#     geom_density_ridges(stat = "density_ridges",
#                         jittered_points = F,
#                         position = position_points_jitter(width = 0.05, height = 0),
#                         point_shape = '|', point_size = 3, point_alpha = .5, alpha = 0.7,) +
#     scale_fill_cyclical(name = "Color scheme",
#                         values = c("orange", "gray")) +
#     theme(axis.text.x = element_text(angle = 0),
#           axis.ticks.length.x = unit(20, "pt"))+
#     geom_vline(xintercept = median_nb, 
#                col = 2) +
#     scale_color_cyclical(name = "Color scheme",
#                          values = c("black", "black")) +
#     scale_x_log10(breaks = c(1, 10, 20, 50, 100, 500, 1000, median_nb), 
#                   labels = c("1","10","20","50","100","500", "1000", median_nb)) +
#     theme_ridges() +
#     xlab("Array sizes") +
#     ylab("Number of arrays") +
#     ggtitle(projectname) 
#   
#   plt
#   
# }
```

```{r}
# arraySize_tbl_grp <- function(input_tibble, projectname=""){
#   ggthemr_reset()
#     
#   d1 <- input_tibble %>% 
#     group_by(SRA_ID) %>% 
#     count(`# spacers`) %>% 
#     mutate(nspacers = `# spacers`) %>% 
#     select(-`# spacers`)
# 
#   d2 <- d1 %>%
#     split(list(.$SRA_ID)) %>%
#     purrr::map(.y = names(.),
#          ~ ggplot(data=., aes(x=nspacers, y=n)) +
#            geom_bar(stat='identity', width = 1)+
#            facet_grid(.~SRA_ID) +
#            labs(title = .y) +
#            ggtitle(paste(projectname , .$SRA_ID, sep=" - ")))
# 
#   d2
# }

#arraySize_tbl_grp(PRJEB15432_unmerged_stats, "PRJEB15432")
```


```{r}
# #
# # Barplot of the arrays sizes. 
# # Input: PXXXXX_unmerged_stats
# #
# arraySize_tbl <- function(input_tibble, projectname=""){
#   d2 <- input_tibble %>% 
#     mutate(`DR concensus` <- as.factor(`DR concensus`)) %>% 
#     count(`# spacers`) 
#   
#   
#   # Size the x axis labels
#   max <- max(d2)
#   tik = 5
#   if(max > 200){
#     tik = 10
#   }else if(max > 300){
#     tik = 20
#   }
#   
#   # Title string
#   sTitle = paste("Project ", projectname, " - Arrays sizes", sep = "")
#   
#   fig2 <- plot_ly(d2, x = ~`# spacers`, y=~n,  type = 'bar', marker = list(color = "orange"))
#   
#   fig2 <- fig2 %>% 
#     layout(title = sTitle, 
#            yaxis = list(title = 'Array Count'), 
#            xaxis = list(dtick = tik, 
#                         tick0 = 0, 
#                         tickmode = "linear", 
#                         mode = "markers+lines"))
#   fig2
#   
#   
# }

#arraySize_tbl(PRJEB15432_unmerged_stats, "PRJEB15432")
```



## Stats
```{r}
# #
# #
# #
# merged_spacers_stats <- function(input_tibble){
#   # mutate the tibble to have group and spacer name 
#   
#   input_grouped <- input_tibble %>% 
#     mutate(Group = str_extract(names, "G[0-9]+")) %>% 
#     mutate(Spacer = str_extract(names, "SP[0-9]+"))
#   
#   input_grouped %>% 
#     group_by(Group) %>% 
#     summarise(n = n())
#   
#   input_grouped %>%
#     extract(Group, c("prefix", "group_value"), regex = "([g])?(\\d)?") %>% 
#     group_by(group_value)
# 
# }

```

## Number of DR
```{r}
# #
# # Retrieves basic numbers
# # Input: PXXX_unmerged_stats
# # Optional: confirm numbers by using PXXX_merged_stats
# #
# basic_stats <- function(input_tibble, projectname=""){
#   print(input_tibble %>% summarize(`Number of entries (DR): ` = n()))
#   print(input_tibble %>% summarize(`Total spacers: ` = sum(`# spacers`)))
# }
# 
# 
# #basic_stats(PRJEB15432_unmerged_stats, "PRJEB15432")

```


## Number of spacers per million reads

```{r,results='hide', echo=FALSE}
# For each sample, number of spacers per million reads
# Input: unmerged stats


nb_spacers_norm <- function(input_tibble, projectname=""){
  ggthemr_reset()
    
  d1 <- input_tibble %>% 
    mutate(nspacers = `# spacers`) %>% 
    mutate(SRA_ID = as.factor(SRA_ID)) %>% 
    select(-`# spacers`) %>% 
    group_by(SRA_ID) %>% 
    summarise(ProjectID, total_spacers = sum(nspacers), million_reads = reads_count, normalized_total = sum(nspacers)/reads_count, n=n()) %>% 
    distinct(SRA_ID, .keep_all = TRUE) %>% 
    ungroup(SRA_ID)
  
  return(d1)
}




```

## Sample w/wo CRISPR

```{r,results='hide', echo=FALSE}
#
# Plots of the number of samples with/out CRISPR 
# Input: PXXX_unmerged_ids
#
samples_crispr_presence <- function(input_tibble, projectname){
  
  asdf <- input_tibble %>% 
    mutate(no_crispr = !no_crispr) %>% 
    group_by(no_crispr) %>% 
    summarise(n = n())
  
  
  fig <- asdf %>% 
    ggplot(aes(y = no_crispr, x = n, fill = no_crispr)) + 
    geom_col(color="black") +
    scale_fill_manual(values=c("grey", "orange")) + 
    geom_text(aes(label=n ), hjust = 0, nudge_x = -2) +
    labs(title = paste("Presence of CRISPR in the samples in project ", 
                       projectname, 
                       sep = ""), 
         y="CRISPR present", 
         x = "Number of samples") + 
    labs(fill = "Contains CRISPR") + 
    theme(legend.title = element_text(colour = "grey"))
  
  show(fig)
}

#samples_crispr_presence(PRJEB15432_unmerged_ids, "PRJEB15432")

```



# Tests TBR
Attention: now all functions will be executed from the full merged dataset at the end of the file.
```{r,results='hide', echo=FALSE}
# # Inputs
# 
# # ---- Merged (sratools) ----
# 
# PRJEB6952_merged_data <- get_bioproject_tibbles("PRJEB6952", "../4_CRASS/")
# 
# PRJEB6952_merged_spacers  <- PRJEB6952_merged_data$spacers
# PRJEB6952_merged_dr <- PRJEB6952_merged_data$dr
# PRJEB6952_merged_stats <- PRJEB6952_merged_data$stats
# 
# 
# arraySize_tbl(PRJEB6952_merged_stats, "PRJEB6952")
# 
# # ---- Not merged (sratools) ----
# 
# PRJEB6952_all_unmerged_data <- get_sra_tibbles("PRJEB6952", "../4_CRASS/")
# 
# PRJEB6952_unmerged_spacers  <- PRJEB6952_all_unmerged_data$spacers
# PRJEB6952_unmerged_dr <- PRJEB6952_all_unmerged_data$dr
# PRJEB6952_unmerged_stats <- PRJEB6952_all_unmerged_data$stats
# PRJEB6952_unmerged_ids <- PRJEB6952_all_unmerged_data$id_tbl
# 
# 
# arraySize_tbl(PRJEB6952_unmerged_stats, "PRJEB6952")
# arraySize_tbl_grp_ridge(PRJEB6952_unmerged_stats)


```


```{r,results='hide', echo=FALSE}
# #
# 
# # ---- Merged (sratools) ----
# 
# PRJEB15432_merged_data <- get_bioproject_tibbles("PRJEB15432", "../4_CRASS/")
# 
# PRJEB15432_merged_spacers  <- PRJEB15432_merged_data$spacers #
# PRJEB15432_merged_dr <- PRJEB15432_merged_data$dr#
# PRJEB15432_merged_stats <- PRJEB15432_merged_data$stats#
# 
# arraySize_tbl(PRJEB15432_merged_stats, "PRJEB15432")
# 
# 
# 
# # ---- Not merged (sratools) ----
# 
# PRJEB15432_all_unmerged_data <- get_sra_tibbles("PRJEB15432", "../4_CRASS/")
# 
# PRJEB15432_unmerged_spacers  <- PRJEB15432_all_unmerged_data$spacers
# PRJEB15432_unmerged_dr <- PRJEB15432_all_unmerged_data$dr
# PRJEB15432_unmerged_stats <- PRJEB15432_all_unmerged_data$stats
# PRJEB15432_unmerged_ids <- PRJEB15432_all_unmerged_data$id_tbl # IDs of SRA entries with zero CRISPRs 
# 
# 
# 
# arraySize_tbl(PRJEB15432_unmerged_stats, "PRJEB15432")
# 
# arraySize_tbl_grp_ridge(PRJEB15432_unmerged_stats)
# 
# PRJEB15432_nb_spacers_norm = nb_spacers_norm(PRJEB15432_unmerged_stats)
# 
# PRJEB15432_nb_spacers_norm_plot = nb_spacers_norm_plot(PRJEB15432_unmerged_stats, "PRJEB15432")
# 
# samples_crispr_presence(PRJEB15432_unmerged_ids, "PRJEB15432")
# 
# # TODO: merge (left join) the arraySize_grp_ridge to the Nb_spacers_norm to have the ones with 0 spacers as well
# 
# # NOTE: The arrays with the unmerged and the merged don't match: 
# #       setdiff(PRJEB15432_unmerged_stats$`DR concensus` ,PRJEB15432_merged_stats$`DR concensus` )
# #       View(PRJEB15432_unmerged_stats[PRJEB15432_unmerged_stats$`DR concensus` == "TCACGTGATCAGTGTATGATCAGCCACGTGA",])
# #       One DR hasn't been merged, no idea why. 


```

TODO: mean line overall / ticks at the mean
 X axis logscale, interest in the first 100
 -> coord_trans(y="log2")



```{r,results='hide', echo=FALSE}
# tsts


#tbldna <- dss2tibble(dna)

# 
# asdf <- PRJEB6952_merged_spacers %>% 
#     mutate(Group = str_extract(names, "G[0-9]+")) %>% 
#     mutate(Spacer = str_extract(names, "SP[0-9]+"))
# 
# 
# 
# asdf %>% 
#   group_by(Group) %>% 
#   summarise(n = n())
# 
# asdf %>%
#   extract(Group, "group_value", regex = "(\\d)") %>% 
#   extract(Spacer, "spacer_value", regex = "(\\d+)") %>% 
#   group_by(group_value)
# 
# 
# 

# stats

#merged_spacers_stats("PRJEB6952", "../4_CRASS/")

# Size

#arraySize()






```





#-----------------------------------

# BioProjects analysis

#-----------------------------------

# Main dataframes

```{r}
spacers_data <- tibble()
repeats_data <- tibble()
stats_data <- tibble()
id_data <- tibble()
matched_cas_data <- tibble()
```



#-----------------------------------

## xPRJEB6952

Pinking defect in cheese
Thermus thermophilus is responsible for the pink discolouration defect in cheese


https://www.ncbi.nlm.nih.gov/bioproject/PRJEB6952


```{r ,results='hide', message=FALSE, echo=FALSE, warning=FALSE}
# Inputs

# projectname = "PRJEB6952"
# location = "../4_CRASS/"
# 
# # ---- Project data ----
# 
# project_data <- get_sra_tibbles(projectname, location)
# project_matched_cas <- get_matched_cas_CCFDB(projectname, location)
# 
# 
# 
# 
# 
# 
# # ---- Merge to main dataframes ---- 
# 
# 
# spacers_data = bind_rows(spacers_data, project_data$spacers)
# repeats_data = bind_rows(repeats_data, project_data$dr)
# stats_data = bind_rows(stats_data, project_data$stats)
# id_data = bind_rows(id_data, project_data$id_tbl)
# matched_cas_data = bind_rows(matched_cas_data, project_matched_cas)
# 


```






<!-- #----------------------------------- -->
<!-- ## PRJEB15423 -->

<!-- Description -->

<!-- Omics-based insights into flavor development and microbial succession within surface-ripened cheese -->

<!-- https://www.ncbi.nlm.nih.gov/bioproject/PRJEB15423 -->





#-----------------------------------
## xPRJEB15432

Description

Microbial succession and flavour production in the fermented dairy beverage kefir

https://www.ncbi.nlm.nih.gov/bioproject/PRJEB15432


Data 

```{r ,results='hide', message=FALSE, echo=FALSE, warning=FALSE}


# Inputs
# 
# projectname = "PRJEB15432"
# location = "../4_CRASS/"
# 
# 
# project_data <- get_sra_tibbles(projectname, location)
# 
# # ---- Cas matches ----
# project_matched_cas <- get_matched_cas_CCFDB(projectname, location)
# 
# 
# # ---- Merge to main dataframes ---- 
# 
# 
# spacers_data = bind_rows(spacers_data, project_data$spacers)
# repeats_data = bind_rows(repeats_data, project_data$dr)
# stats_data = bind_rows(stats_data, project_data$stats)
# id_data = bind_rows(id_data, project_data$id_tbl)
# matched_cas_data = bind_rows(matched_cas_data, project_matched_cas)


```




#-----------------------------------
## PRJEB32768

Cotter lab CS study

https://www.ncbi.nlm.nih.gov/bioproject/PRJEB32768
https://www.nature.com/articles/s43016-020-0129-3

Manually removed entry 3340807 -> empty folder generates error.

```{r ,results='hide', message=FALSE, echo=FALSE, warning=FALSE}

# Inputs

projectname = "PRJEB32768"
location = "../4_CRASS/"


project_data <- get_sra_tibbles(projectname, location)

# ---- Cas matches ----
project_matched_cas <- get_matched_cas_CCFDB(projectname, location)


# ---- Merge to main dataframes ---- 


spacers_data = bind_rows(spacers_data, project_data$spacers)
repeats_data = bind_rows(repeats_data, project_data$dr)
stats_data = bind_rows(stats_data, project_data$stats)
id_data = bind_rows(id_data, project_data$id_tbl)
matched_cas_data = bind_rows(matched_cas_data, project_matched_cas)

```




#-----------------------------------
## xPRJEB35321

Cotter Lab FF study

https://www.ncbi.nlm.nih.gov/bioproject/PRJEB35321
https://msystems.asm.org/content/5/6/e00522-20

Fermented Food Study TEAGASC - Irish Agronomic Food research. 


```{r ,results='hide', message=FALSE, echo=FALSE, warning=FALSE}
# Inputs

# projectname = "PRJEB35321"
# location = "../4_CRASS/"
# 
# 
# project_data <- get_sra_tibbles(projectname, location)
# 
# # ---- Cas matches ----
# project_matched_cas <- get_matched_cas_CCFDB(projectname, location)
# 
# 
# # ---- Merge to main dataframes ---- 
# 
# 
# spacers_data = bind_rows(spacers_data, project_data$spacers)
# repeats_data = bind_rows(repeats_data, project_data$dr)
# stats_data = bind_rows(stats_data, project_data$stats)
# id_data = bind_rows(id_data, project_data$id_tbl)
# matched_cas_data = bind_rows(matched_cas_data, project_matched_cas)

```



#-----------------------------------
## PRJNA286900

https://www.ncbi.nlm.nih.gov/bioproject/286900

Cotija cheese is a Mexican dairy product of spontaneous fermentation with a particular sensorial profile and an indisputable microbiological quality. In this study we analysed bacterial Cotija cheese metagenome in order to get a complete picture of bacterial diversity as well as metabolic potential of consortium related to flavour and odour compound production as well as gene related with bacteriocins production.

```{r ,results='hide', message=FALSE, echo=FALSE, warning=FALSE}
# Inputs

projectname = "PRJNA286900"
location = "../4_CRASS/"


project_data <- get_sra_tibbles(projectname, location)

# ---- Cas matches ----
project_matched_cas <- get_matched_cas_CCFDB(projectname, location)


# ---- Merge to main dataframes ---- 


spacers_data = bind_rows(spacers_data, project_data$spacers)
repeats_data = bind_rows(repeats_data, project_data$dr)
stats_data = bind_rows(stats_data, project_data$stats)
id_data = bind_rows(id_data, project_data$id_tbl)
matched_cas_data = bind_rows(matched_cas_data, project_matched_cas)
```




#-----------------------------------
## xPRJNA430402 


E.coli O157 and E.coli O121 spiked in cheddar cheese
Evaluation of Enriched Background Microflora of Raw Milk Cheese Spiked with E. coli O157:H7 and E. coli O103 using Next-Generation Sequencing Technology

https://www.ncbi.nlm.nih.gov/bioproject/PRJNA430402


```{r ,results='hide', message=FALSE, echo=FALSE, warning=FALSE}
# Inputs

# projectname = "PRJNA430402"
# location = "../4_CRASS/"
# 
# 
# project_data <- get_sra_tibbles(projectname, location)
# 
# # ---- Cas matches ----
# project_matched_cas <- get_matched_cas_CCFDB(projectname, location)
# 
# 
# # ---- Merge to main dataframes ---- 
# 
# 
# spacers_data = bind_rows(spacers_data, project_data$spacers)
# repeats_data = bind_rows(repeats_data, project_data$dr)
# stats_data = bind_rows(stats_data, project_data$stats)
# id_data = bind_rows(id_data, project_data$id_tbl)
# matched_cas_data = bind_rows(matched_cas_data, project_matched_cas)
```




#-----------------------------------
## xPRJNA482503  

https://www.ncbi.nlm.nih.gov/bioproject/PRJNA482503

The role of Parmesan cheese in vectoring bovine microbiotaâ€™s engraftment in the human gut.

```{r ,results='hide', message=FALSE, echo=FALSE, warning=FALSE}
# # Inputs
# 
# projectname = "PRJNA482503"
# location = "../4_CRASS/"
# 
# 
# project_data <- get_sra_tibbles(projectname, location)
# 
# # ---- Cas matches ----
# project_matched_cas <- get_matched_cas_CCFDB(projectname, location)
# 
# 
# # ---- Merge to main dataframes ---- 
# 
# 
# spacers_data = bind_rows(spacers_data, project_data$spacers)
# repeats_data = bind_rows(repeats_data, project_data$dr)
# stats_data = bind_rows(stats_data, project_data$stats)
# id_data = bind_rows(id_data, project_data$id_tbl)
# matched_cas_data = bind_rows(matched_cas_data, project_matched_cas)
```


#-----------------------------------
## PRJNA603575

https://www.ncbi.nlm.nih.gov/bioproject/?term=PRJNA603575

food metagenome
Yogurt and dietary supplement metagenomes


```{r ,results='hide', message=FALSE, echo=FALSE, warning=FALSE}
# Inputs

projectname = "PRJNA603575"
location = "../4_CRASS/"


project_data <- get_sra_tibbles(projectname, location)

# ---- Cas matches ----
project_matched_cas <- get_matched_cas_CCFDB(projectname, location)


# ---- Merge to main dataframes ---- 


spacers_data = bind_rows(spacers_data, project_data$spacers)
repeats_data = bind_rows(repeats_data, project_data$dr)
stats_data = bind_rows(stats_data, project_data$stats)
id_data = bind_rows(id_data, project_data$id_tbl)
matched_cas_data = bind_rows(matched_cas_data, project_matched_cas)
```

#-----------------------------------
## PRJEB23938

https://www.ncbi.nlm.nih.gov/bioproject/PRJEB23938

Metagenomic and Metatranscriptomic Analysis of the Microbial Community in Swiss-type Maasdam Cheese During Ripening

ONLY METAGENOMIC SAMPLES WERE KEPT


```{r ,results='hide', message=FALSE, echo=FALSE, warning=FALSE, eval=FALSE}
# Inputs

projectname = "PRJEB23938"
location = "../4_CRASS/"


project_data <- get_sra_tibbles(projectname, location)

# ---- Cas matches ----
project_matched_cas <- get_matched_cas_CCFDB(projectname, location)


# ---- Merge to main dataframes ---- 


spacers_data = bind_rows(spacers_data, project_data$spacers)
repeats_data = bind_rows(repeats_data, project_data$dr)
stats_data = bind_rows(stats_data, project_data$stats)
id_data = bind_rows(id_data, project_data$id_tbl)
matched_cas_data = bind_rows(matched_cas_data, project_matched_cas)
```

#-----------------------------------
## PRJEB30079

https://www.ncbi.nlm.nih.gov/bioproject/PRJEB30079


```{r ,results='hide', message=FALSE, echo=FALSE, warning=FALSE, eval=FALSE}
# Inputs

projectname = "PRJEB30079"
location = "../4_CRASS/"


project_data <- get_sra_tibbles(projectname, location)

# ---- Cas matches ----
project_matched_cas <- get_matched_cas_CCFDB(projectname, location)


# ---- Merge to main dataframes ---- 


spacers_data = bind_rows(spacers_data, project_data$spacers)
repeats_data = bind_rows(repeats_data, project_data$dr)
stats_data = bind_rows(stats_data, project_data$stats)
id_data = bind_rows(id_data, project_data$id_tbl)
matched_cas_data = bind_rows(matched_cas_data, project_matched_cas)
```


#-----------------------------------
## PRJ




```{r ,results='hide', message=FALSE, echo=FALSE, warning=FALSE}
# # Inputs
# 
# projectname = ""
# location = "../4_CRASS/"
# 
# 
# project_data <- get_sra_tibbles(projectname, location)
# 
# # ---- Cas matches ----
# project_matched_cas <- get_matched_cas_CCFDB(projectname, location)
# 
# 
# # ---- Merge to main dataframes ---- 
# 
# 
# spacers_data = bind_rows(spacers_data, project_data$spacers)
# repeats_data = bind_rows(repeats_data, project_data$dr)
# stats_data = bind_rows(stats_data, project_data$stats)
# id_data = bind_rows(id_data, project_data$id_tbl)
# matched_cas_data = bind_rows(matched_cas_data, project_matched_cas)
```


#-------------------------
## 20201217_metagenomes

Hard cheese starter cultures from gruyere, emmentaler, sbrinz
Agroscope



```{r ,results='hide', message=FALSE, echo=FALSE, warning=FALSE}
# Inputs

projectname = "20201217_metagenomes"
location = "../4_CRASS/"

# ---- Project data ----

project_data <- get_sra_tibbles(projectname, location)
project_matched_cas <- get_matched_cas_CCFDB(projectname, location)






# ---- Merge to main dataframes ---- 


spacers_data = bind_rows(spacers_data, project_data$spacers)
repeats_data = bind_rows(repeats_data, project_data$dr)
stats_data = bind_rows(stats_data, project_data$stats)
id_data = bind_rows(id_data, project_data$id_tbl)
matched_cas_data = bind_rows(matched_cas_data, project_matched_cas)



```




#-----------------------------------

# Join DR to spacers 

Join the direct repeats to the spacers.  

```{r ,results='hide', message=FALSE, echo=FALSE, warning=FALSE}
CRISPRscope_meta_tbl <- inner_join(spacers_data, repeats_data, by = c("ProjectID", "SRA_ID", "GID")) %>% 
  mutate(spacer_length = length.x, repeat_length = length.y, read_count = reads_count.x) %>% 
  select(-reads_count.y, -names.y, -length.x, -length.y, -reads_count.x)
head(CRISPRscope_meta_tbl)
```

#-----------------------------------

# Outputs
## DR - SP - DR

Create a fasta file containing the repeat - spacers - repeat sequences for each spacer to be mapped to the reads. The goal here is to see the proportion of spacer reads that actually might be from the protospacers. 

ATTENTION: change the chunk's eval argument to execute

```{r, eval=TRUE}
#
# For each project - sample - group: 
#     create DR - SP - DR sequences uniquely named
#

out1 <- CRISPRscope_meta_tbl %>% 
  group_by(ProjectID, SRA_ID, GID) %>% 
  mutate(header = paste(ProjectID, SRA_ID, GID, SPID, Coverage, sep = "_") , 
         seq = paste(repeat_seq, spacer_seq, repeat_seq, sep = "")) %>% 
  ungroup %>% 
  select(header, seq) 


df <- data.frame(out1$header, out1$seq)
df.fasta = dataframe2fas(df, file="./OUTPUT/CRISPRscope_DR_SP_DR.fasta")


```

```{r}
# Trying to avoid duplicate sequences

# out1 <- CRISPRscope_meta_tbl %>% 
#   group_by(ProjectID) %>% 
#   distinct(ProjectID, spacer_seq) %>% 
#   rowid_to_column("id") %>% 
#   unite(header, c("ProjectID", "id"), sep = "_") %>% 
#   left_join(CRISPRscope_meta_tbl, by = "spacer_seq") %>% 
#   mutate(header = paste(ProjectID, SRA_ID, GID, SPID, Coverage, sep = "_") , 
#          seq = paste(repeat_seq, spacer_seq, repeat_seq, sep = "")) %>% 
#   ungroup %>% 
#   select(header, seq) 
# 
# 
# df <- data.frame(out1$header, out1$seq)
# df.fasta = dataframe2fas(df, file="CRISPRscope_DR_SP_DR_distinct.fasta")
```



## Export Spacers fasta



```{r, eval = TRUE}


out2 <- CRISPRscope_meta_tbl %>%
  group_by(ProjectID, SRA_ID, GID) %>%
  mutate(header = paste(ProjectID, SRA_ID, GID, SPID, Coverage, sep = "_") ,
         seq = spacer_seq) %>%
  ungroup %>%
  select(header, seq)

dfs <- data.frame(out2$header, out2$seq)
dfs.fasta = dataframe2fas(dfs, file = "./OUTPUT/CRISPRscope_meta_SP.fasta")


```

## Export Repeats 


```{r}

repeats_unique <- repeats_data %>% select(repeat_seq) %>% unique() 

out_repeat <- repeats_data %>% 
  mutate(header = paste(ProjectID, SRA_ID, GID, names, sep = "_"), seq = repeat_seq) %>%
  select(header, seq) %>% 
  distinct(seq, .keep_all = TRUE)

dfs <- data.frame(out_repeat$header, out_repeat$seq)
dfs.fasta = dataframe2fas(dfs, file="./OUTPUT/CRISPRscope_meta_DR.fasta")


write.table(repeats_unique, "./OUTPUT/CRISPRscope_meta_unique_repeats.tsv", na = "", quote = FALSE, sep = "\t",row.names = FALSE, col.names = FALSE)

```

Repeats were processed on https://cctyper.crispr.dk/#/submit
Check: https://www.liebertpub.com/doi/abs/10.1089/crispr.2020.0059



#-----------------------------------
# LOOP BACK Import CRISPRCasTyper

```{r}
CRISPRscope_meta_cctyper_cas_types <- read_table2("C:/Users/thsch/Desktop/master_thesis/1_scripts/INPUT/CRISPRscope_meta_cctyper_cas-types_v2.tsv") %>%
  mutate(Prediction_probability = Prediction) %>% 
  select(-probability, -Prediction)
#View(CRISPRscope_meta_cctyper_cas_types)
```


#-----------------------------------
# LOOP BACK Join cas type
Join CCT cas type and prediction probability to 


```{r}
# Join the CRISPRscope_meta_cctyper_cas_types to the CRISPRscope_meta_tbl 

CRISPRscope_meta_tbl <- left_join(CRISPRscope_meta_tbl, CRISPRscope_meta_cctyper_cas_types, by = c("repeat_seq" = "Sequence"))
CRISPRscope_meta_tbl

```
#-----------------------------------

# LOOP BACK Import clusters 

```{r}
Clusters_CRISPRscope_meta_spacers <- read_csv("C:/Users/thsch/Desktop/master_thesis/1_scripts/INPUT/Clusters_CRISPRscope_meta_SP.csv") %>% 
  distinct(seq, cluster, identity)

Clusters_CRISPRscope_meta_repeats <- read_csv("C:/Users/thsch/Desktop/master_thesis/1_scripts/INPUT/Clusters_CRISPRscope_meta_DR.csv") %>% 
  distinct(seq, cluster, identity)


```


# Join cluster info

```{r}
CRISPRscope_meta_tbl <- CRISPRscope_meta_tbl %>% 
  left_join(Clusters_CRISPRscope_meta_spacers, by = c("spacer_seq" = "seq")) %>% 
  mutate(cluster_spacer = cluster) %>% 
  mutate(identity_spacer_cluster = identity) %>% 
  select(-cluster, -identity) %>% 
  relocate(cluster_spacer, .after = "spacer_seq") %>% 
  relocate(identity_spacer_cluster, .after = "cluster_spacer")

CRISPRscope_meta_tbl <- CRISPRscope_meta_tbl %>% 
  left_join(Clusters_CRISPRscope_meta_repeats, by = c("repeat_seq" = "seq")) %>% 
  mutate(cluster_repeat = cluster) %>% 
  mutate(identity_repeat_cluster = identity) %>% 
  select(-cluster, -identity) %>% 
  relocate(cluster_repeat, .after = "repeat_seq") %>% 
  relocate(identity_repeat_cluster, .after = "cluster_repeat")


```



#-----------------------------------



# Available datasets


```{r}
spacers_data 
repeats_data 
stats_data 
id_data 
matched_cas_data 

# CRISPRCasTyper - Repeat -> cas type identification
CRISPRscope_meta_cctyper_cas_types

# Repeats added to spacers_data
CRISPRscope_meta_tbl

CRISPRscope_meta_tbl %>% summarise(n_dr = n_distinct(spacer_seq))

```

#-----------------------------------
# Export Datasets

```{r}
# CSV format is heavier but usable everywhere
write.csv(CRISPRscope_meta_tbl, file = "./OUTPUT/CRISPRscope_meta.csv")


# Save an object to a file in RDS format is less space-consuming
saveRDS(CRISPRscope_meta_tbl, file = "CRISPRscope_meta_tbl.rds")
```


### Export bioproject and sample list

```{r}

CRISPRscope_meta_tbl %>% select(ProjectID, SRA_ID) %>% distinct() %>% write.csv(file = "./OUTPUT/CRISPRscope_meta_Project_Sample_list.csv", col.names = TRUE)

CRISPRscope_meta_tbl %>% select(ProjectID) %>% distinct()

```


#-----------------------------------
#-----------------------------------


# Plots


## Ridge array size TODO
```{r fig.height=20, fig.width=25}


#TODO: take the 1 array out and write this in methods


# Cleaning and grouping

tmp <- stats_data %>%
  group_by(ProjectID, SRA_ID) %>%
  count(`# spacers`) %>%
  mutate(nspacers = `# spacers`) %>%
  select(-`# spacers`)

# %>%  
#   filter(n() > 1) %>% 
#   ungroup




for (proj in unique(stats_data$ProjectID)) {
  
  
  print(proj)
  
  l <- tmp %>% filter(ProjectID == proj) 
  
  
  
  median_nb <- median(l$nspacers)
  print(median_nb)

  plt <- l %>% ggplot(aes(
    y = SRA_ID,
    x = nspacers,
    fill = SRA_ID,
    color = SRA_ID)) +
    geom_density_ridges(
      stat = "binline",
      #stat = "density_ridges",
      jittered_points = F,
      position = position_points_jitter(width = 0.05, height = 0),
      point_shape = '|',
      point_size = 3,
      point_alpha = .5,
      alpha = 0.7,
    ) +
    scale_fill_cyclical(name = "Color scheme",
                        values = c("orange", "gray")) +
    theme(axis.text.x = element_text(angle = 0),
          axis.ticks.length.x = unit(20, "pt")) +
    geom_vline(xintercept = median_nb,
               col = 2) +
    scale_color_cyclical(name = "Color scheme",
                         values = c("black", "black")) +
    scale_x_log10(
      breaks = c(1, 10, 20, 50, 100, 500, 1000, median_nb),
      labels = c("1", "10", "20", "50", "100", "500", "1000", median_nb)
    ) +
    theme_ridges() +
    xlab("Array sizes") +
    ylab("Number of arrays") +
    ggtitle(proj)
  
  show(plt)
  print("done")
  
}
```

## Array size barplot TODO

```{r  fig.height=20, fig.width=40}
#
# Barplot of the arrays sizes. 
# Input: PXXXXX_unmerged_stats
#stats_data

d2 <- stats_data  %>% 
  mutate(`DR concensus` <- as.factor(`DR concensus`)) %>%
  group_by(ProjectID) %>%
  count(`# spacers`) %>% 
  ggplot(aes(x=`# spacers`, y=n)) +
  geom_col(fill = "orange") +
  labs(title = "CRISPR arrays size and number",
       x = "Number of spacers",
       y = "Array count") +
  theme(plot.title = element_text(color="black", size=20, face="bold"),
        axis.title.x = element_text(color="black", size=16),
        axis.title.y = element_text(color="black", size=16)) +
  facet_wrap(~ProjectID, scales = "free")

ggsave("./IMG/metagenome/array_size_barplot.png", plot = d2, height = 20, width = 40)

d2



#0 => for the one with 2000+ spacers: View(CRISPRscope_meta_tbl %>% filter(ProjectID == "PRJNA482503") %>% count(SRA_ID, GID))
# result: G4
# View(CRISPRscope_meta_tbl %>% filter(ProjectID == "PRJNA482503") %>% filter(GID == "G4"))


```

## Second option
```{r}
# Size the x axis labels

tmp <- stats_data  %>% 
  mutate(`DR concensus` <- as.factor(`DR concensus`)) %>%
  group_by(ProjectID) %>%
  count(`# spacers`)

for (proj in unique(stats_data$ProjectID)) {
  print(proj)
  
  d2 <- tmp %>% filter(ProjectID == proj)
  
  
  max <- max(d2$`# spacers`)
  tik = 5
  if (max > 200) {
    tik = 10
  } else if (max > 300) {
    tik = 20
  }
  
  # Title string
  sTitle = paste("Project ", proj, " - Arrays sizes", sep = "")
  
  fig2 <-
    plot_ly(
      d2,
      x = ~ `# spacers`,
      y =  ~ n,
      type = 'bar',
      marker = list(color = "orange")
    )
  
  fig2 <- fig2 %>%
    layout(
      title = sTitle,
      yaxis = list(title = 'Array Count'),
      xaxis = list(
        title = "Number of spacers",
        dtick = tik,
        tick0 = 0,
        tickmode = "linear",
        mode = "markers+lines"
      )
    ) %>%
  layout(autosize = T, margin=list( l = 50, r = 50, b = 100, t = 100,  pad = 4))
  
  
  show(fig2)
  
}

```
 


## CRISPR content TODO

```{r}
#
# Plots of the number of samples with/out CRISPR 
# Input: PXXX_unmerged_ids
#
for (proj in unique(id_data$ProjectID)) {
  
  
  print(proj)

  asdf <- id_data %>%
    mutate(no_crispr = !no_crispr) %>%
    group_by(no_crispr) %>%
    filter(ProjectID == proj) %>% 
    summarise(n = n())
  
  fig <- asdf %>% ungroup %>% 
    ggplot(aes(y = no_crispr, x = n, fill = no_crispr)) +
    geom_col(color = "black") +
    scale_fill_manual(values = c("grey", "orange")) +
    geom_text(aes(label = n), hjust = 0, nudge_x = -2) +
    labs(
      title = paste("Presence of CRISPR in the samples in project ",
                    proj,
                    sep = ""),
      y = "CRISPR present",
      x = "Number of samples"
    ) 
  show(fig)
}
```

#*report*
```{r fig.height=8, fig.width=15}
id_data = id_data %>% 
    mutate(ProjectID = replace(ProjectID, ProjectID == "20201217_metagenomes", "Agroscope"))

plt <- id_data %>% mutate(no_crispr = !no_crispr) %>% 
  count(ProjectID, no_crispr) %>% 
  ggplot(aes(x=ProjectID, y=n, fill=no_crispr)) +
  labs(x = "Projects", y = "Sample count") +
  theme_bw() +
  theme(axis.text.x = element_text(size = 8, angle = 45), axis.title.x = element_text(size = 8),
        axis.text.y = element_text(size = 8), axis.title.y = element_text(size = 8),
        plot.title = element_text(size = 15, face = "bold", color = "darkgreen")) +
  geom_col(position="stack") +
  scale_fill_discrete(name = "Presence of \nCRISPR") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust=1, color = "black"),
        axis.text.y = element_text(color = "black"),
        axis.title.x = element_text(color = "black"),
        axis.title.y = element_text(color = "black"),
        #plot.title = element_text(color = "grey"),
        panel.grid.major = element_line(size = 0.25, linetype = 'solid',colour = "black"),
        panel.grid.minor = element_line(size = 0.25, linetype = 'solid',colour = "black"),
        panel.background = element_rect(fill = "white",colour = "black", linetype = 'solid', size = 0.5),
        plot.background = element_rect(fill = "white",colour = NA),
        plot.margin = unit(c(1,1,1,1), "cm"),
        legend.position = "right",
        legend.background = element_rect(fill = "white", colour = NA),
        legend.box.background = element_rect(fill = "white", colour = NA),
        legend.text = element_text(color = "black"),
        legend.title = element_text(color = "black")) 

ggsave(
  plot = plt,
  file = "./IMG/report/BioProject_CRISPR_content.pdf",
  bg = "white",
  width = 25,
  height = 15,
  units = "cm",
  dpi = 800
)
  
plt # %>% layout(title ="CRISPR content", annotations=list(yref='paper',xref="paper",y=1.05,x=1.15, text="Presence of CRISPR",showarrow=F) )
```


## Number of Spacers / mil. reads

```{r  fig.height=20, fig.width=25}

# https://stackoverflow.com/questions/38833042/r-making-bigger-graphs-when-using-facet-wrap

#
# Per sample, number of spacers per million reads
# input: unmerged stats
# Call: nb_spacers_norm

# TODO: to match all the graphs, maybe order by the same as the "number of arrays " graph.

nb_sp_nrm <- nb_spacers_norm(stats_data) %>%
  arrange(normalized_total) %>%
  mutate(SRA_ID = fct_reorder(SRA_ID, normalized_total))


fig2 <- nb_sp_nrm %>% mutate(SRA_ID = fct_reorder(SRA_ID, normalized_total)) %>%
  ggplot(aes(x=as_factor(SRA_ID), y=normalized_total)) +
  geom_bar(stat="identity", fill="orange", alpha=.6, width=.5)+
  ylab("Number of spacers per million reads") +
  ylim(0, 0.0005)+
  xlab("Sample") +
  ggtitle("Number of spacers per million reads") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
  facet_wrap(~ProjectID, scales='fixed') 

fig2

#fig2 <- ggplotly(fig2)# %>% layout(autosize = F, width = 2000, height = 1500, margin = 3)

#fig2



```
For vincent's data only
```{r  fig.height=20, fig.width=25}

# https://stackoverflow.com/questions/38833042/r-making-bigger-graphs-when-using-facet-wrap

#
# Per sample, number of spacers per million reads
# input: unmerged stats
# Call: nb_spacers_norm

# TODO: to match all the graphs, maybe order by the same as the "number of arrays " graph.

nb_sp_nrm <- nb_spacers_norm(stats_data) %>%
  arrange(normalized_total) %>%
  mutate(SRA_ID = fct_reorder(SRA_ID, normalized_total))


fig2 <- nb_sp_nrm %>% filter(ProjectID == "20201217_metagenomes") %>% mutate(SRA_ID = fct_reorder(SRA_ID, normalized_total)) %>%
  ggplot(aes(x=as_factor(SRA_ID), y=normalized_total)) +
  geom_bar(stat="identity", fill="orange", alpha=.6, width=.5)+
  ylab("Number of spacers per million reads") +
  ylim(0, 0.00005)+
  xlab("Sample") +
  ggtitle("Number of spacers per million reads") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust=1, color = "grey", size = 10),
        axis.text.y = element_text(color = "grey"),
        axis.title.x = element_text(color = "grey"),
        axis.title.y = element_text(color = "grey"),
        plot.title = element_text(color = "grey"),
        panel.grid.major = element_line(size = 0.25, linetype = 'solid',colour = "grey"),
        panel.grid.minor = element_line(size = 0.25, linetype = 'solid',colour = "grey"),
        panel.background = element_rect(fill = "transparent",colour = "grey", linetype = 'solid', size = 0.5),
        plot.background = element_rect(fill = "transparent",colour = NA),
        plot.margin = unit(c(1,1,1,1), "cm"),
        legend.position = "none") 

ggsave(
  plot = fig2,
  file = "./IMG/metagenome/TRANS_Vincent_richness.png",
  type = "cairo-png",
  bg = "transparent",
  width = 65,
  height = 35,
  units = "cm",
  dpi = 800
)


fig2

#fig2 <- ggplotly(fig2)# %>% layout(autosize = F, width = 2000, height = 1500, margin = 3)

#fig2



```






# Between-projects comparison


Dairy:
PRJEB6952
PRJEB32768
PRJNA286900
PRJNA430402

Various fermented food:
PRJEB35321
PRJNA603575



Unclear:
PRJNA482503  16S rRNA gene profiling Parmesan cheese production chain





## Cas-Subtypes per projects
#*report*
```{r}
#View(CRISPRscope_meta_tbl)
plt_castype <- CRISPRscope_meta_tbl %>%
  mutate(ProjectID = replace(ProjectID, ProjectID == "20201217_metagenomes", "Agroscope")) %>%
  count(ProjectID, Subtype, GID) %>% count(Subtype, ProjectID) %>% 
  select(ProjectID, Subtype, n) %>%
  ggplot(aes(x = ProjectID, y = Subtype, fill = n)) +
  geom_tile(color = "gray") +
  scale_fill_gradient(low = "light blue", high = "green") +
  theme_bw() +
  theme(
    axis.text.x = element_text(
      angle = 45,
      vjust = 1,
      hjust = 1,
      color = "black"
    ),
    axis.text.y = element_text(hjust = 0, color = "black"),
    axis.title.x = element_text(color = "black"),
    axis.title.y = element_text(color = "black"),
    plot.title = element_text(color = "black"),
    axis.line.y.left = element_line(color = "black"),
    axis.line.y.right = element_line(color = "black"),
    axis.line.x.top = element_line(color = "black"),
    axis.line.x.bottom = element_line(color = "black"),
    axis.ticks = element_line(color = "black"),
    panel.background = element_rect(fill = "white", colour = NA),
    panel.grid.major = element_line(
      size = 0,
      linetype = 'solid',
      colour = "white"
    ),
    panel.grid.minor = element_line(
      size = 0,
      linetype = 'solid',
      colour = "white"
    ),
    plot.background = element_rect(fill = "white", colour = NA),
    plot.margin = unit(c(1, 1, 1, 1), "cm"),
    legend.position = "right",
    legend.background = element_rect(fill = "white", colour = NA),
    legend.box.background = element_rect(fill = "white", colour = NA),
    legend.text = element_text(color = "black"),
    legend.title = element_text(color = "black")
  ) +
  labs(fill = "Array count") +
  xlab("Bioproject") +
  ylab("Cas-subtype")


ggsave(
  plot = plt_castype,
  file = "./IMG/report/Cas-Subtypes_bioproject.pdf",
  bg = "white",
  width = 25,
  height = 15,
  units = "cm",
  dpi = 800
)

plt_castype
```



# matched Organism analysis

```{r}
matched_cas_data %>% distinct(DirectRepeat)
```























