2\_CRISPRscope\_metagenomic\_data
================
Thibault Schowing
13 février, 2022

-   [Parsing of the CRASS outputs](#parsing-of-the-crass-outputs)
-   [Various functions](#various-functions)
    -   [Imports](#imports)
    -   [Directories](#directories)
    -   [Utils](#utils)
-   [Data getters](#data-getters)
    -   [Get SRA entries](#get-sra-entries)
    -   [TBR Matched Direct Repeats –&gt;](#tbr-matched-direct-repeats-)
-   [Data analysis functions](#data-analysis-functions)
    -   [Number of spacers per million
        reads](#number-of-spacers-per-million-reads)
    -   [Sample w/wo CRISPR](#sample-wwo-crispr)
-   [Main dataframes](#main-dataframes)
-   [Join DR to spacers](#join-dr-to-spacers)
-   [Low-high quality samples
    content](#low-high-quality-samples-content)
-   [Import clusters](#import-clusters)
-   [Cluster 80%](#cluster-80)
-   [Cluster 100%](#cluster-100)
-   [Sample filtering](#sample-filtering)
-   [Available datasets](#available-datasets)
    -   [Full Datasets](#full-datasets)
    -   [DR - SP - DR](#dr---sp---dr)
    -   [Export Spacers](#export-spacers)
    -   [Export Repeats](#export-repeats)

# Parsing of the CRASS outputs

For each metagenome assembly with CRASS, various outputs are produced. -
spacers.fasta - DR.fasta - flankers.fasta - stats.csv

The stats.csv gives useful informations on the general quantity of
CRISPR arrays and their sizes.

# Various functions

## Imports

    ## -- Attaching packages --------------------------------------- tidyverse 1.3.1 --

    ## v ggplot2 3.3.5     v dplyr   1.0.7
    ## v tibble  3.1.2     v stringr 1.4.0
    ## v tidyr   1.1.3     v forcats 0.5.1
    ## v purrr   0.3.4

    ## -- Conflicts ------------------------------------------ tidyverse_conflicts() --
    ## x dplyr::filter() masks stats::filter()
    ## x dplyr::lag()    masks stats::lag()

    ## 
    ## Attachement du package : 'plotly'

    ## L'objet suivant est masqué depuis 'package:ggplot2':
    ## 
    ##     last_plot

    ## L'objet suivant est masqué depuis 'package:stats':
    ## 
    ##     filter

    ## L'objet suivant est masqué depuis 'package:graphics':
    ## 
    ##     layout

    ## Le chargement a nécessité le package : BiocGenerics

    ## Le chargement a nécessité le package : parallel

    ## 
    ## Attachement du package : 'BiocGenerics'

    ## Les objets suivants sont masqués depuis 'package:parallel':
    ## 
    ##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
    ##     clusterExport, clusterMap, parApply, parCapply, parLapply,
    ##     parLapplyLB, parRapply, parSapply, parSapplyLB

    ## Les objets suivants sont masqués depuis 'package:dplyr':
    ## 
    ##     combine, intersect, setdiff, union

    ## Les objets suivants sont masqués depuis 'package:stats':
    ## 
    ##     IQR, mad, sd, var, xtabs

    ## Les objets suivants sont masqués depuis 'package:base':
    ## 
    ##     anyDuplicated, append, as.data.frame, basename, cbind, colnames,
    ##     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
    ##     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
    ##     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
    ##     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
    ##     union, unique, unsplit, which.max, which.min

    ## Le chargement a nécessité le package : S4Vectors

    ## Le chargement a nécessité le package : stats4

    ## 
    ## Attachement du package : 'S4Vectors'

    ## L'objet suivant est masqué depuis 'package:plotly':
    ## 
    ##     rename

    ## Les objets suivants sont masqués depuis 'package:dplyr':
    ## 
    ##     first, rename

    ## L'objet suivant est masqué depuis 'package:tidyr':
    ## 
    ##     expand

    ## Les objets suivants sont masqués depuis 'package:base':
    ## 
    ##     expand.grid, I, unname

    ## Le chargement a nécessité le package : IRanges

    ## 
    ## Attachement du package : 'IRanges'

    ## L'objet suivant est masqué depuis 'package:plotly':
    ## 
    ##     slice

    ## Les objets suivants sont masqués depuis 'package:dplyr':
    ## 
    ##     collapse, desc, slice

    ## L'objet suivant est masqué depuis 'package:purrr':
    ## 
    ##     reduce

    ## L'objet suivant est masqué depuis 'package:grDevices':
    ## 
    ##     windows

    ## Le chargement a nécessité le package : XVector

    ## 
    ## Attachement du package : 'XVector'

    ## L'objet suivant est masqué depuis 'package:purrr':
    ## 
    ##     compact

    ## Le chargement a nécessité le package : GenomeInfoDb

    ## 
    ## Attachement du package : 'Biostrings'

    ## L'objet suivant est masqué depuis 'package:base':
    ## 
    ##     strsplit

## Directories

``` r
repository <- "C:/Users/thsch/Desktop/CRISPRscope" # Main git repository. 
google_drive_folder <- "C:/Users/thsch/Google Drive/0_Documents/1_Emploi/E_StageAgroscope/CRISPRscope_writing/IMG"
data_folder <- "C:/Users/thsch/Desktop/0_data"
```

## Utils

# Data getters

## Get SRA entries

## TBR Matched Direct Repeats –&gt;

Get the repeats and corresponding organisms that were found in the
CRISPRCasFinderDB.

# Data analysis functions

## Number of spacers per million reads

## Sample w/wo CRISPR

\#———————————–

\#———————————–

# Main dataframes

``` r
spacers_data <- tibble()
repeats_data <- tibble()
stats_data <- tibble()
id_data <- tibble()
matched_cas_data <- tibble()

sample_data <- tibble()
```

\#———————————– \#\# PRJEB32768

Cotter lab CS study

<https://www.ncbi.nlm.nih.gov/bioproject/PRJEB32768>
<https://www.nature.com/articles/s43016-020-0129-3>

Manually removed entry 3340807 -&gt; empty folder generates error.

\#———————————– \#\# PRJNA286900

<https://www.ncbi.nlm.nih.gov/bioproject/286900>

Cotija cheese is a Mexican dairy product of spontaneous fermentation
with a particular sensorial profile and an indisputable microbiological
quality. In this study we analysed bacterial Cotija cheese metagenome in
order to get a complete picture of bacterial diversity as well as
metabolic potential of consortium related to flavour and odour compound
production as well as gene related with bacteriocins production.

\#———————————– \#\# PRJNA603575

<https://www.ncbi.nlm.nih.gov/bioproject/?term=PRJNA603575>

food metagenome Yogurt and dietary supplement metagenomes

\#———————————– \#\# PRJEB23938

<https://www.ncbi.nlm.nih.gov/bioproject/PRJEB23938>

Metagenomic and Metatranscriptomic Analysis of the Microbial Community
in Swiss-type Maasdam Cheese During Ripening

ONLY METAGENOMIC SAMPLES WERE KEPT

Manually removed sample ERR2227971 -&gt; empty

Actually doesn’t contain any CRISPR, =&gt; empty stat file causes error,
sample ID manually added in “low quality sample” section

\#———————————– \#\# PRJEB30079

<https://www.ncbi.nlm.nih.gov/bioproject/PRJEB30079>

The antithrombotic properties of raw and pasteurised goat milk and
cheeses Food Research International

\#———————————– \#\# CheeseRaclette

\#————————- \#\# 20201217\_metagenomes

Hard cheese starter cultures from gruyere, emmentaler, sbrinz Agroscope

\#———————————–

# Join DR to spacers

Join the direct repeats to the spacers.

\#———————————– \# Quality filtering

``` r
CRISPRscope_meta_tbl_raw <- CRISPRscope_meta_tbl

CRISPRscope_meta_tbl <- CRISPRscope_meta_tbl_raw %>% 
  filter(Coverage > 1) %>%                            # Remove low coverage
  filter(str_length(spacer_seq) > 15) %>%             # Remove spurious spacers
  filter(str_length(spacer_seq) < 50) %>%             # Remove huge spacers (Munch et al)
  filter(spacer_seq %>% str_detect("^[ACTGactg]+$")) %>%  # Originally, no spacer / repeat contain N's Must be a particularity of CRISPRCasFinder
  filter(repeat_seq %>% str_detect("^[ACTGactg]+$"))



# CRISPRscope_meta_tbl %>% filter(ProjectID == "CheeseRaclette") %>% distinct(SRA_ID)
```

# Low-high quality samples content

mark low quality samples in id\_data as samples without CRISPR

``` r
id_data  #=> 203 rows 
```

    ## # A tibble: 203 x 4
    ##    ProjectID  SRA_ID     reads_count no_crispr
    ##    <chr>      <chr>            <dbl> <lgl>    
    ##  1 PRJEB32768 ERR3340746     9021920 FALSE    
    ##  2 PRJEB32768 ERR3340747     8980400 FALSE    
    ##  3 PRJEB32768 ERR3340748     8134426 FALSE    
    ##  4 PRJEB32768 ERR3340749     9105958 FALSE    
    ##  5 PRJEB32768 ERR3340750     6212746 FALSE    
    ##  6 PRJEB32768 ERR3340751     8519550 FALSE    
    ##  7 PRJEB32768 ERR3340752     7947418 FALSE    
    ##  8 PRJEB32768 ERR3340753     6489248 FALSE    
    ##  9 PRJEB32768 ERR3340754     8642442 FALSE    
    ## 10 PRJEB32768 ERR3340755     7885652 FALSE    
    ## # ... with 193 more rows

``` r
sample_data
```

    ## # A tibble: 203 x 2
    ##    sra_sample ProjectID 
    ##    <chr>      <chr>     
    ##  1 ERR3340746 PRJEB32768
    ##  2 ERR3340747 PRJEB32768
    ##  3 ERR3340748 PRJEB32768
    ##  4 ERR3340749 PRJEB32768
    ##  5 ERR3340750 PRJEB32768
    ##  6 ERR3340751 PRJEB32768
    ##  7 ERR3340752 PRJEB32768
    ##  8 ERR3340753 PRJEB32768
    ##  9 ERR3340754 PRJEB32768
    ## 10 ERR3340755 PRJEB32768
    ## # ... with 193 more rows

``` r
All_samples = sample_data

# list of samples where good spacers are
CRISPRscope_meta_high_quality <- CRISPRscope_meta_tbl %>% select(SRA_ID) %>% distinct()

# elements present in "low quality" but not in "high" have to be added to id_data with no_crispr = true

low_quality_or_empty = All_samples %>% 
  filter(!sra_sample %in%  CRISPRscope_meta_high_quality$SRA_ID ) %>% 
  add_row(sra_sample = "ERR2226381", ProjectID = "PRJEB23938") %>% 
  add_row(sra_sample = "ERR2226382", ProjectID = "PRJEB23938") %>% 
  add_row(sra_sample = "ERR2226383", ProjectID = "PRJEB23938") %>% 
  add_row(sra_sample = "ERR2226384", ProjectID = "PRJEB23938") %>% 
  add_row(sra_sample = "ERR2226385", ProjectID = "PRJEB23938") %>% 
  add_row(sra_sample = "ERR2226386", ProjectID = "PRJEB23938") %>% 
  add_column(CRISPR = FALSE) %>% 
  mutate(SRA_ID = sra_sample) %>% select(-sra_sample) %>% 
  relocate(SRA_ID, .after = "ProjectID") %>% 
  relocate(CRISPR, .after = "SRA_ID")


high_quality = CRISPRscope_meta_tbl %>% select(SRA_ID, ProjectID) %>% 
  distinct() %>% 
  add_column(CRISPR = TRUE)%>% 
  relocate(SRA_ID, .after = "ProjectID") %>% 
  relocate(CRISPR, .after = "SRA_ID")
 
# no sample are in both          
#intersect(low_quality_or_empty$sra_sample, high_quality$SRA_ID)

# Merge the two df, use in CRISPR content barchart

sample_crispr_content <- bind_rows(low_quality_or_empty, high_quality) %>% write.csv(file = paste(data_folder, "/IMPORT_EXPORT/CRISPRscope_meta_Project_Sample_list.csv", sep=""), col.names = TRUE)
```

    ## Warning in write.csv(., file = paste(data_folder, "/IMPORT_EXPORT/
    ## CRISPRscope_meta_Project_Sample_list.csv", : attempt to set 'col.names' ignored

# Import clusters

    ## Rows: 17016 Columns: 4

    ## -- Column specification --------------------------------------------------------
    ## Delimiter: ","
    ## chr (2): seq, seqid
    ## dbl (2): cluster, identity

    ## 
    ## i Use `spec()` to retrieve the full column specification for this data.
    ## i Specify the column types or set `show_col_types = FALSE` to quiet this message.

    ## Rows: 556 Columns: 4

    ## -- Column specification --------------------------------------------------------
    ## Delimiter: ","
    ## chr (2): seq, seqid
    ## dbl (2): cluster, identity

    ## 
    ## i Use `spec()` to retrieve the full column specification for this data.
    ## i Specify the column types or set `show_col_types = FALSE` to quiet this message.

# Cluster 80%

Remove old clustering info for spacers and repeats and add new ones
(global, done with all repeats and spacers)

``` r
CRISPRscope_meta_tbl <- CRISPRscope_meta_tbl %>% 
  left_join(Clusters_CRISPRscope_meta_spacers, by = c("spacer_seq" = "seq")) %>% 
  mutate(cluster_spacer = cluster) %>% 
  mutate(identity_spacer_cluster = identity) %>% 
  select(-cluster, -identity) %>% 
  relocate(cluster_spacer, .after = "spacer_seq") %>% 
  relocate(identity_spacer_cluster, .after = "cluster_spacer")

CRISPRscope_meta_tbl <- CRISPRscope_meta_tbl %>% 
  left_join(Clusters_CRISPRscope_meta_repeats, by = c("repeat_seq" = "seq")) %>% 
  mutate(cluster_repeat = cluster) %>% 
  mutate(identity_repeat_cluster = identity) %>% 
  select(-cluster, -identity) %>% 
  relocate(cluster_repeat, .after = "repeat_seq") %>% 
  relocate(identity_repeat_cluster, .after = "cluster_repeat")
```

# Cluster 100%

Note that the clustering was added before removing the spacers bigger
than 72bp. The clustering file is then containing additional potentially
unused clusters.

``` r
Clusters_CRISPRscope_spacers_identity <- read_csv(paste(data_folder, "/IMPORT/clustering/Clusters1_CRISPRscope_ALL_SPACERS.csv", sep="")) %>% 
  distinct(seq, cluster, identity) %>% select(-identity)
```

    ## Rows: 79007 Columns: 4

    ## -- Column specification --------------------------------------------------------
    ## Delimiter: ","
    ## chr (2): seq, seqid
    ## dbl (2): cluster, identity

    ## 
    ## i Use `spec()` to retrieve the full column specification for this data.
    ## i Specify the column types or set `show_col_types = FALSE` to quiet this message.

``` r
Clusters_CRISPRscope_repeats_identity <- read_csv(paste(data_folder, "/IMPORT/clustering/Clusters1_CRISPRscope_ALL_REPEATS.csv", sep="")) %>% 
  distinct(seq, cluster, identity) %>% select(-identity)
```

    ## Rows: 1362 Columns: 4

    ## -- Column specification --------------------------------------------------------
    ## Delimiter: ","
    ## chr (2): seq, seqid
    ## dbl (2): cluster, identity

    ## 
    ## i Use `spec()` to retrieve the full column specification for this data.
    ## i Specify the column types or set `show_col_types = FALSE` to quiet this message.

``` r
CRISPRscope_meta_tbl <- CRISPRscope_meta_tbl %>% 
  left_join(Clusters_CRISPRscope_spacers_identity, by = c("spacer_seq" = "seq")) %>% 
  mutate(cluster_spacer_identity = cluster) %>% 
  select(-cluster) %>% 
  relocate(cluster_spacer_identity, .after = "cluster_spacer")


CRISPRscope_meta_tbl <- CRISPRscope_meta_tbl %>% 
  left_join(Clusters_CRISPRscope_repeats_identity, by = c("repeat_seq" = "seq")) %>% 
  mutate(cluster_repeat_identity = cluster) %>% 
  select(-cluster) %>% 
  relocate(cluster_repeat_identity, .after = "cluster_repeat")
```

\#———————————– \# Spacers per milion reads

``` r
CRISPRscope_meta_tbl <- CRISPRscope_meta_tbl %>%
  group_by(SRA_ID) %>%
  summarise(SRA_ID, total_spacers = n_distinct(cluster_spacer_identity)) %>% 
  distinct() %>% 
  left_join(CRISPRscope_meta_tbl, by = c("SRA_ID")) %>% 
  relocate(SRA_ID, .after="ProjectID") %>% 
  relocate(total_spacers, .after="read_count") %>% 
  group_by(SRA_ID) %>% 
  mutate(spacer_per_milread = (total_spacers/read_count)*1000000) %>% 
  relocate(spacer_per_milread, .after="read_count") %>% 
  ungroup()
```

    ## `summarise()` has grouped output by 'SRA_ID'. You can override using the `.groups` argument.

\#———————————– \# Repeats per milion reads

``` r
CRISPRscope_meta_tbl <- CRISPRscope_meta_tbl %>%
  group_by(SRA_ID) %>%
  summarise(SRA_ID, total_repeats = n_distinct(cluster_repeat_identity)) %>% 
  distinct() %>% 
  left_join(CRISPRscope_meta_tbl, by = c("SRA_ID")) %>% 
  relocate(SRA_ID, .after="ProjectID") %>% 
  relocate(total_repeats, .after="read_count") %>% 
  group_by(SRA_ID) %>% 
  mutate(repeat_per_milread = (total_repeats/read_count)*1000000) %>% 
  relocate(repeat_per_milread, .after="read_count") %>% 
  ungroup()
```

    ## `summarise()` has grouped output by 'SRA_ID'. You can override using the `.groups` argument.

\#———————————– \# LOOP BACK Import CRISPRCasTyper results

Use the exported repeat tsv file to infer the cas subtypes using
CRISPRCasTyper -&gt; Repeat\_typer tool

``` r
CRISPRscope_meta_cctyper_cas_types <- read_table2(paste(data_folder, "/IMPORT/cctyper/CRISPRscope_meta_cctyper_cas-types.tsv", sep="")) %>%
  mutate(Prediction_probability = Prediction) %>% 
  select(-probability, -Prediction)
```

    ## Warning: `read_table2()` was deprecated in readr 2.0.0.
    ## Please use `read_table()` instead.

    ## 
    ## -- Column specification --------------------------------------------------------
    ## cols(
    ##   Sequence = col_character(),
    ##   Subtype = col_character(),
    ##   Prediction = col_double(),
    ##   probability = col_character()
    ## )

    ## Warning: 556 parsing failures.
    ## row col  expected    actual                                                                                  file
    ##   1  -- 4 columns 3 columns 'C:/Users/thsch/Desktop/0_data/IMPORT/cctyper/CRISPRscope_meta_cctyper_cas-types.tsv'
    ##   2  -- 4 columns 3 columns 'C:/Users/thsch/Desktop/0_data/IMPORT/cctyper/CRISPRscope_meta_cctyper_cas-types.tsv'
    ##   3  -- 4 columns 3 columns 'C:/Users/thsch/Desktop/0_data/IMPORT/cctyper/CRISPRscope_meta_cctyper_cas-types.tsv'
    ##   4  -- 4 columns 3 columns 'C:/Users/thsch/Desktop/0_data/IMPORT/cctyper/CRISPRscope_meta_cctyper_cas-types.tsv'
    ##   5  -- 4 columns 3 columns 'C:/Users/thsch/Desktop/0_data/IMPORT/cctyper/CRISPRscope_meta_cctyper_cas-types.tsv'
    ## ... ... ......... ......... .....................................................................................
    ## See problems(...) for more details.

``` r
#View(CRISPRscope_meta_cctyper_cas_types)

#CRISPRscope_meta_cctyper_cas_types %>% summarise(ndis_seq = n_distinct(Sequence), nseq = n())


# Join the CRISPRscope_meta_cctyper_cas_types to the CRISPRscope_meta_tbl 

CRISPRscope_meta_tbl$repeat_seq <- as.factor(CRISPRscope_meta_tbl$repeat_seq)
CRISPRscope_meta_cctyper_cas_types$Sequence <- as.factor(CRISPRscope_meta_cctyper_cas_types$Sequence)


CRISPRscope_meta_tbl <- left_join(CRISPRscope_meta_tbl, CRISPRscope_meta_cctyper_cas_types, by = c("repeat_seq" = "Sequence"))
CRISPRscope_meta_tbl
```

    ## # A tibble: 25,626 x 22
    ##    ProjectID    SRA_ID GID   SPID  Coverage spacer_seq            cluster_spacer
    ##    <chr>        <chr>  <chr> <chr> <chr>    <chr>                          <dbl>
    ##  1 CheeseRacle~ 1944   G36   SP451 7        TACGTTCTTTGAACCAAGCT~           1676
    ##  2 CheeseRacle~ 1944   G36   SP6   10       GTGGATATTTTGAACTATTG~           1677
    ##  3 CheeseRacle~ 1944   G36   SP11  34       GTAGCTCTCTTAATGTTGTC~            544
    ##  4 CheeseRacle~ 1944   G36   SP16  58       GATAGATTCCTGGCAATTCT~           1678
    ##  5 CheeseRacle~ 1944   G36   SP13~ 2        TTTCTAATTCAATCATCAGC~           2263
    ##  6 CheeseRacle~ 1944   G36   SP19  45       TACCGTCTAACAAATTCAAA~           2264
    ##  7 CheeseRacle~ 1944   G36   SP153 16       TTCATTTCCATCATGCTTCT~           1679
    ##  8 CheeseRacle~ 1944   G36   SP186 34       CAAATTTTAAAGGCTGTAAA~           1124
    ##  9 CheeseRacle~ 1944   G36   SP850 2        CGATAAGGTCAACAGAATGT~            411
    ## 10 CheeseRacle~ 1944   G36   SP26  21       AGGCATTGCCTGAGAGATTA~            412
    ## # ... with 25,616 more rows, and 15 more variables:
    ## #   cluster_spacer_identity <dbl>, identity_spacer_cluster <dbl>,
    ## #   repeat_seq <fct>, cluster_repeat <dbl>, cluster_repeat_identity <dbl>,
    ## #   identity_repeat_cluster <dbl>, spacer_length <int>, repeat_length <int>,
    ## #   read_count <dbl>, repeat_per_milread <dbl>, total_repeats <int>,
    ## #   spacer_per_milread <dbl>, total_spacers <int>, Subtype <chr>,
    ## #   Prediction_probability <dbl>

\#———————————–

\#———————————–

# Sample filtering

Refine samples according to metadata. Here we remove the samples taken
specifically from the rind of cheese instead of core.

``` r
CRISPRscope_meta_METADATA <- read_delim(paste(data_folder, "/IMPORT/metadata/CRISPRscope_meta_Project_Sample_MANUAL_METADATA_VS.csv", sep=""),"\t", escape_double = FALSE, trim_ws = TRUE)
```

    ## New names:
    ## * `` -> ...1

    ## Rows: 209 Columns: 10

    ## -- Column specification --------------------------------------------------------
    ## Delimiter: "\t"
    ## chr (8): ProjectID, SRA_ID, BioSample_ID, Cheese_type, Sample_location (core...
    ## dbl (1): ...1
    ## lgl (1): CRISPR

    ## 
    ## i Use `spec()` to retrieve the full column specification for this data.
    ## i Specify the column types or set `show_col_types = FALSE` to quiet this message.

``` r
# Get the samples from Rind
rind_list <- CRISPRscope_meta_METADATA  %>% filter(`Sample_location (core rind)` == "Rind") %>% select(ProjectID, SRA_ID)
# => 22 samples from PRJEB32768

# Remove samples from main dataset according to metadata
CRISPRscope_meta_tbl_filtered <- CRISPRscope_meta_tbl %>% filter(!SRA_ID %in% rind_list$SRA_ID)
```

\#———————————–

# Available datasets

``` r
# NO QUALITY FILTERING ON THESE
# spacers_data 
# repeats_data 
# stats_data 
id_data 
```

    ## # A tibble: 203 x 4
    ##    ProjectID  SRA_ID     reads_count no_crispr
    ##    <chr>      <chr>            <dbl> <lgl>    
    ##  1 PRJEB32768 ERR3340746     9021920 FALSE    
    ##  2 PRJEB32768 ERR3340747     8980400 FALSE    
    ##  3 PRJEB32768 ERR3340748     8134426 FALSE    
    ##  4 PRJEB32768 ERR3340749     9105958 FALSE    
    ##  5 PRJEB32768 ERR3340750     6212746 FALSE    
    ##  6 PRJEB32768 ERR3340751     8519550 FALSE    
    ##  7 PRJEB32768 ERR3340752     7947418 FALSE    
    ##  8 PRJEB32768 ERR3340753     6489248 FALSE    
    ##  9 PRJEB32768 ERR3340754     8642442 FALSE    
    ## 10 PRJEB32768 ERR3340755     7885652 FALSE    
    ## # ... with 193 more rows

``` r
# matched_cas_data 

# CRISPRCasTyper - Repeat -> cas type identification
CRISPRscope_meta_cctyper_cas_types
```

    ## # A tibble: 556 x 3
    ##    Sequence                                Subtype Prediction_probability
    ##    <fct>                                   <chr>                    <dbl>
    ##  1 AAATACAACCGCTCCTCGATAAAAGGGGACGAGAACT   III-A                    0.987
    ##  2 GATCTATTTTAGCTGAAAACTGAAGGAATCAATAGC    II-C                     0.966
    ##  3 GTTGAACTACTCATTGATTTGATACTCTTCTAAAAC    II-A                     0.999
    ##  4 AGGATCACCCCCGCATGTGCGGGTAATAC           I-E                      0.999
    ##  5 ATTTACATTCCACTATTTTTAGATTTGTAC          I-B                      0.996
    ##  6 GTTGTACAGTTACTTAAATCTTGAGAGTACAAAAAC    II-C                     0.528
    ##  7 GGGTTTAACCTTATTGATTTAACATCCTTCTAAAAC    II-A                     0.998
    ##  8 ATTTCAATTCACGCACTCATATAGAGTGCGAC        I-C                      0.971
    ##  9 GCTTTAGAATCTTGGTAAATCAATAGGGTCAAGAACTTA II-A                     0.975
    ## 10 CGGTTTATCCCCGCTGGCGCGGGGAACAC           I-E                      1.00 
    ## # ... with 546 more rows

``` r
# Repeats added to spacers_data
CRISPRscope_meta_tbl
```

    ## # A tibble: 25,626 x 22
    ##    ProjectID    SRA_ID GID   SPID  Coverage spacer_seq            cluster_spacer
    ##    <chr>        <chr>  <chr> <chr> <chr>    <chr>                          <dbl>
    ##  1 CheeseRacle~ 1944   G36   SP451 7        TACGTTCTTTGAACCAAGCT~           1676
    ##  2 CheeseRacle~ 1944   G36   SP6   10       GTGGATATTTTGAACTATTG~           1677
    ##  3 CheeseRacle~ 1944   G36   SP11  34       GTAGCTCTCTTAATGTTGTC~            544
    ##  4 CheeseRacle~ 1944   G36   SP16  58       GATAGATTCCTGGCAATTCT~           1678
    ##  5 CheeseRacle~ 1944   G36   SP13~ 2        TTTCTAATTCAATCATCAGC~           2263
    ##  6 CheeseRacle~ 1944   G36   SP19  45       TACCGTCTAACAAATTCAAA~           2264
    ##  7 CheeseRacle~ 1944   G36   SP153 16       TTCATTTCCATCATGCTTCT~           1679
    ##  8 CheeseRacle~ 1944   G36   SP186 34       CAAATTTTAAAGGCTGTAAA~           1124
    ##  9 CheeseRacle~ 1944   G36   SP850 2        CGATAAGGTCAACAGAATGT~            411
    ## 10 CheeseRacle~ 1944   G36   SP26  21       AGGCATTGCCTGAGAGATTA~            412
    ## # ... with 25,616 more rows, and 15 more variables:
    ## #   cluster_spacer_identity <dbl>, identity_spacer_cluster <dbl>,
    ## #   repeat_seq <fct>, cluster_repeat <dbl>, cluster_repeat_identity <dbl>,
    ## #   identity_repeat_cluster <dbl>, spacer_length <int>, repeat_length <int>,
    ## #   read_count <dbl>, repeat_per_milread <dbl>, total_repeats <int>,
    ## #   spacer_per_milread <dbl>, total_spacers <int>, Subtype <chr>,
    ## #   Prediction_probability <dbl>

``` r
CRISPRscope_meta_tbl_filtered
```

    ## # A tibble: 24,111 x 22
    ##    ProjectID    SRA_ID GID   SPID  Coverage spacer_seq            cluster_spacer
    ##    <chr>        <chr>  <chr> <chr> <chr>    <chr>                          <dbl>
    ##  1 CheeseRacle~ 1944   G36   SP451 7        TACGTTCTTTGAACCAAGCT~           1676
    ##  2 CheeseRacle~ 1944   G36   SP6   10       GTGGATATTTTGAACTATTG~           1677
    ##  3 CheeseRacle~ 1944   G36   SP11  34       GTAGCTCTCTTAATGTTGTC~            544
    ##  4 CheeseRacle~ 1944   G36   SP16  58       GATAGATTCCTGGCAATTCT~           1678
    ##  5 CheeseRacle~ 1944   G36   SP13~ 2        TTTCTAATTCAATCATCAGC~           2263
    ##  6 CheeseRacle~ 1944   G36   SP19  45       TACCGTCTAACAAATTCAAA~           2264
    ##  7 CheeseRacle~ 1944   G36   SP153 16       TTCATTTCCATCATGCTTCT~           1679
    ##  8 CheeseRacle~ 1944   G36   SP186 34       CAAATTTTAAAGGCTGTAAA~           1124
    ##  9 CheeseRacle~ 1944   G36   SP850 2        CGATAAGGTCAACAGAATGT~            411
    ## 10 CheeseRacle~ 1944   G36   SP26  21       AGGCATTGCCTGAGAGATTA~            412
    ## # ... with 24,101 more rows, and 15 more variables:
    ## #   cluster_spacer_identity <dbl>, identity_spacer_cluster <dbl>,
    ## #   repeat_seq <fct>, cluster_repeat <dbl>, cluster_repeat_identity <dbl>,
    ## #   identity_repeat_cluster <dbl>, spacer_length <int>, repeat_length <int>,
    ## #   read_count <dbl>, repeat_per_milread <dbl>, total_repeats <int>,
    ## #   spacer_per_milread <dbl>, total_spacers <int>, Subtype <chr>,
    ## #   Prediction_probability <dbl>

``` r
CRISPRscope_meta_tbl %>% summarise(n_dr = n_distinct(spacer_seq))
```

    ## # A tibble: 1 x 1
    ##    n_dr
    ##   <int>
    ## 1 10669

\#———————————– \# Export

## Full Datasets

``` r
saveRDS(id_data, file = paste(data_folder, "/CRISPRscope_meta_results/id_data.rds", sep=""))


# CSV format is heavier but usable everywhere
write.csv(CRISPRscope_meta_tbl, file = paste(data_folder, "/CRISPRscope_meta_results/CRISPRscope_meta_tbl.csv", sep=""))


# Save an object to a file in RDS format is less space-consuming
saveRDS(CRISPRscope_meta_tbl, file = paste(data_folder, "/CRISPRscope_meta_results/CRISPRscope_meta_tbl.rds", sep=""))



# CSV format is heavier but usable everywhere
write.csv(CRISPRscope_meta_tbl_filtered, file = paste(data_folder, "/CRISPRscope_meta_results/CRISPRscope_meta_tbl_filtered.csv", sep=""))


# Save an object to a file in RDS format is less space-consuming
saveRDS(CRISPRscope_meta_tbl_filtered, file = paste(data_folder, "/CRISPRscope_meta_results/CRISPRscope_meta_tbl_filtered.rds", sep=""))
```

## DR - SP - DR

Create a fasta file containing the repeat - spacers - repeat sequences
for each spacer to be mapped to the reads. The goal here is to see the
proportion of spacer reads that actually might be from the protospacers.

``` r
#
# For each project - sample - group: 
#     create DR - SP - DR sequences uniquely named
#

out1 <- CRISPRscope_meta_tbl %>% 
  group_by(ProjectID, SRA_ID, GID) %>% 
  mutate(header = paste(ProjectID, SRA_ID, GID, SPID, Coverage, sep = "_") , 
         seq = paste(repeat_seq, spacer_seq, repeat_seq, sep = "")) %>% 
  ungroup %>% 
  select(header, seq) 


df <- data.frame(out1$header, out1$seq)
df.fasta = dataframe2fas(df, file=paste(data_folder, "/EXPORT/fasta/CRISPRscope_DR_SP_DR.fasta", sep=""))
```

## Export Spacers

``` r
out2 <- CRISPRscope_meta_tbl_filtered %>%
  group_by(ProjectID, SRA_ID, GID) %>%
  mutate(header = paste(ProjectID, SRA_ID, GID, SPID, Coverage, sep = "_") ,
         seq = spacer_seq) %>%
  ungroup %>%
  select(header, seq)

dfs <- data.frame(out2$header, out2$seq)
dfs.fasta = dataframe2fas(dfs, file = paste(data_folder, "/EXPORT/fasta/CRISPRscope_meta_spacers.fasta", sep=""))
```

## Export Repeats

``` r
repeats_unique <- CRISPRscope_meta_tbl_filtered %>% select(repeat_seq) %>% distinct(repeat_seq)
repeats_unique <- repeats_unique$repeat_seq

out_repeat <- CRISPRscope_meta_tbl_filtered %>% 
  mutate(header = paste(ProjectID, SRA_ID, GID, sep = "_"), seq = repeat_seq) %>%
  select(header, seq) %>% 
  distinct(seq, .keep_all = TRUE)

dfs <- data.frame(out_repeat$header, out_repeat$seq)
dfs.fasta = dataframe2fas(dfs, file=paste(data_folder, "/EXPORT/fasta/CRISPRscope_meta_repeats.fasta", sep=""))


# For CRISPR cas typer (Copy paste with \n )
write.table(repeats_unique, paste(data_folder, "/EXPORT/CRISPRscope_meta_unique_repeats.tsv", sep=""), na = "", quote = FALSE, sep = "\t",row.names = FALSE, col.names = FALSE)
```

Repeats were processed on <https://cctyper.crispr.dk/#/submit> Check:
<https://www.liebertpub.com/doi/abs/10.1089/crispr.2020.0059>
